---
title: "Analysis of DNA quality as an effect of bead beating settings"
author: "Thomas Bygh Nymann Jensen"
datae: "`r format(Sys.time(), '%d-%m-%Y')`, Aalborg, Denmark"
output: 
  bookdown::html_document2:
    toc: true
---

# This is code to replicate the analysis used for my HT paper
# The data includes Nanodrop, Qubit, ScreenTape, Extaction and Library concentrations
# as well as amplicon V1-4 data

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  warning = FALSE)

# Load the needed packages.
source("load_data.R")
library(patchwork)
library(data.table)
```

Data removed for RPM = 1000 as results are bad - bead beating does not seem to work at this speed.

# Amplicon analysis

## QC - PCR controls

```{r}
# subset pcr-controls (no kit value)
data_controls <- amp_subset_samples(ampvis.data, Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank"))

# evaluate number of reads
tab_reads <- data_controls$metadata %>%
  mutate(NoReads = colSums(data_controls$abund))

# plot number of reads in the two plates with 
plot_controls_reads <- tab_reads %>%
  ggplot(aes(x = reorder(Sample_id, +NoReads),
             y = NoReads, fill = Sample_id)) +
  geom_bar(position = "dodge", stat = "identity", alpha = 1) +
  scale_y_continuous(limits = c(0, 5000), 
                     breaks = seq(0, 5000, 500)) +
  geom_text(data = tab_reads, aes(label = NoReads), 
            position = position_dodge(width = 1), 
            vjust = -0.5, 
            size = 6) +
  labs(
    title = "Number of reads in PCR controls",
    fill = "Control",
    x = "",
    y = "Obtained reads") +
  #scale_fill_manual(values = c("salmon", "royalblue4", "seagreen", "seagreen", "seagreen", "seagreen")) +
  theme_bw(base_size = 14) +
  theme(axis.text.x = element_text(angle = 90))

# plot heatmap of rarefied reads
plot_PCR_POS <- 
  data_PCR %>% 
  amp_subset_samples(Sample_id == "PCR-POS") %>%
  amp_heatmap(
    facet_by = "Soil_type",
    tax_aggregate = "Phylum",
    tax_show = 9,
    color_vector = c("white", "royalblue4"),
    plot_values = TRUE,
    normalise = T
  ) +
  #scale_fill_gradient(low = "white", high = "royalblue4", limits = c(0, 21), breaks = seq(0, 21, 3)) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme(axis.text.x = element_blank(),
        plot.title = element_text("Heatmap of PCR positives")) + 
  theme_bw()

plot_PCR_NEG <-
  data_PCR %>% 
  amp_subset_samples(Sample_id == "PCR-NEG") %>%
  amp_heatmap(
    facet_by = "Soil_type",
    tax_aggregate = "OTU",
    tax_add = "Phylum",
    tax_show = 15,
    color_vector = c("white", "salmon"),
    plot_values = TRUE,
    normalise = T
  ) +
  #scale_fill_gradient(low = "white", high = "salmon", limits = c(0, 70), breaks = seq(0, 70, 10)) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme(axis.text.x = element_blank(),
        plot.title = element_text("Heatmap of PCR negatives")) +
  theme_bw()

# organize plots
plot_PCR_reads / (plot_PCR_NEG + plot_PCR_POS)


# test which OTU's are present in the negative control above 5 reads
neg_subset <- 
  data_PCR %>% 
  amp_subset_samples(Sample_id == "PCR-NEG")

wh_OTU_PCR_NEG <- 
  {neg_subset$abund > 5} %>%
  apply(1, any) %>% 
  {.[.]} %>%
  names()

# evaluate contamination
data_con <- amp_subset_taxa(ampvis.data, tax_vector = wh_OTU_PCR_NEG)

tab_reads_con <- data_con$metadata %>%
  mutate(NoReads = colSums(data_con$abund))

# plot number of reads in the two plates with 
plot_PCR_reads <- tab_reads %>%
  ggplot(aes(x = reorder(Soil_type, +NoReads),
             y = NoReads, fill = Sample_id)) +
  geom_bar(position = "dodge", stat = "identity", alpha = 1) +
  scale_y_continuous(limits = c(0, 5000), 
                     breaks = seq(0, 5000, 500)) +
  geom_text(data = tab_reads, aes(label = NoReads), 
            position = position_dodge(width = 1), 
            vjust = -0.5, 
            size = 6) +
  labs(
    title = "Number of reads in PCR controls",
    fill = "Control",
    x = "",
    y = "Obtained reads") +
  scale_fill_manual(values = c("salmon", "royalblue4")) +
  theme_bw(base_size = 14)

```
Some contamination seems to be present in the negative control. The contaminations seems to be a result of cross-sample or NFW contamination as 26 OTUs are abundant above 5 reads.

## QC - blanks

```{r}
data_blank <- 
  amp_subset_samples(
    dataraw, 
    Soil_type == "Blank" & RPM != "1000")

tab_reads <- 
  data_blank$metadata %>%
  merge(
    data.frame(
      NoReads = apply(
        data_blank$abund, 2, sum)
      ),
    by = 0
  )

# plot number of reads from blanks
plot_blank_reads <-
  tab_reads %>%
  ggplot(
    aes(
      x = RPM,
      y = NoReads
    )
  ) +
  geom_bar(position = "dodge", stat = "identity", alpha = 1, fill = "grey75") +
  scale_y_continuous(limits = c(0, 2200), 
                     breaks = seq(0, 2200, 200)) +
  geom_text(data = tab_reads, 
            aes(label = NoReads), 
            position = position_dodge(width = 1), 
            vjust = -0.5, 
            size = 6) +
  labs(y = "Obtained reads") +
  ggtitle("Number of reads in PCR controls",
    subtitle = "Samples faceted by bead beating intensity") +
  #scale_fill_manual(
  #  values = c("white", "grey75")) +
  theme_bw(base_size = 14)

# plot heatmap of rarefied reads
plot_blank <- 
  data_blank %>% 
  amp_heatmap(
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 9,
    color_vector = c("white", "grey75"),
    plot_values = TRUE,
    normalise = T
  ) +
  #scale_fill_gradient(low = "white", high = "grey75", limits = c(0, 75), breaks = seq(0, 75, 15)) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  ggtitle("Heatmap for blank samples",
    subtitle = "Samples faceted by bead beating intensity"
  ) +
  theme_bw(base_size = 14)

# organize plots
g_blank <- 
  ggarrange(
    plot_blank_reads, 
    plot_blank, 
    ncol = 1)

# save plots as png.
ggsave(
  "Blank_controls.png", 
  plot = g_blank, 
  width = 10, 
  height = 7)
```

Quite a lot of contamination seems to be present in the case of 1200 and 1400 RPM. Most of the OTUs also seem to be present in 1600 and 1800 RPM, but has a much lower read count. 20 OTUs intersect between the contamination in the PCR negatives and the blanks (when n > 5). 

```{r}
# intersecting OTU's
data_blank_sub <- 
  data_blank %>%
  amp_subset_samples(RPM %in% c("1600", "1800"))

wh_OTU_blank <- 
  {
    data_blank_sub$abund > 2
  } %>%
  apply(1, any) %>%
  {
    .[.]
  } %>%
  names()

data_blank_OTU <- 
  data_blank_sub %>%
  amp_subset_taxa(
    tax_vector = wh_OTU_blank)

intersect <- 
  intersect(
    wh_OTU_blank, 
    wh_OTU_PCR_NEG)

data_intersect <- 
  dataraw %>%
  amp_subset_samples(
    !Sample_id %in% c("MFD004-1", "MFD004-3", "MFD004-4", "PCR-POS") & !LibID == "LIB-BLANK-1000-360")

plot_intersect <- 
  data_intersect %>% 
  amp_heatmap(
    facet_by = "LibID",
    tax_aggregate = "OTU",
    tax_add = "Phylum",
    tax_show = 20,
    color_vector = c("white", "grey75"),
    plot_values = TRUE,
    normalise = T
  ) +
  #scale_fill_gradient(low = "white", high = "grey75", limits = c(0, 75), breaks = seq(0, 75, 15)) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  ggtitle("Heatmap for contamination",
    subtitle = "Samples faceted by library ID"
  ) +
  theme_bw(base_size = 14)
```


## QC - samples

```{r}
#data_samples <- 
#  amp_subset_samples(
#   dataraw, 
#    !Soil_type %in% c("Blank", "Filter", "Dry_soil") & !is.na(Kit) & RPM != "1000")

order <- ampvis.data %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank")) %>%
  amp_heatmap(group_by = "Project_id",
              tax_show = 5,
              textmap = TRUE) %>%
  rownames_to_column(var = "rowname") %>%
  pull(rowname)

plot.heat_clay <- ampvis.data %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank")) %>%
  amp_subset_samples(Soil_type == "Clay") %>%
  amp_heatmap(facet_by = "Soil_type",
              group_by = "Beadbeating",
              tax_aggregate = "Phylum",
              min_abundance = 0.1,
              plot_values_size = 3,
              #order_y_by = order,
              color_vector = c("white", "royalblue4"))

plot.heat_sand <- ampvis.data %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank")) %>%
  amp_subset_samples(Soil_type == "Sand") %>%
  amp_heatmap(facet_by = "Soil_type",
              group_by = "Beadbeating",
              tax_aggregate = "Phylum",
              min_abundance = 0.1,
              plot_values_size = 3,
              #order_y_by = order,
              color_vector = c("white", "salmon"))
  #theme(axis.text.x = element_blank())

plot.heat_peat <- ampvis.data %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank")) %>%
  amp_subset_samples(Soil_type == "Peat") %>%
  amp_heatmap(facet_by = "Soil_type",
              group_by = "Beadbeating",
              tax_aggregate = "Phylum",
              min_abundance = 0.1,
              plot_values_size = 3,
              #order_y_by = order,
              color_vector = c("white", "seagreen"))
  #theme(axis.text.x = element_blank())

ampvis.data %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank")) %>%
  amp_rarecurve(color_by = "Soil_type") +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon"))
  
plot.ord_all <- ampvis.data %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank"), rarefy = 5000) %>%
  amp_ordinate(filter_species = 0.1, 
               type = "PCoA", 
               distmeasure = "bray",
               transform = "none",
               species_plot = FALSE,
               sample_color_by = "Soil_type",
               sample_shape_by = "RPM",
               #species_nlabels = 3,
               #species_label_taxonomy = "Genus",
               #sample_plotly = TRUE,
               sample_colorframe = FALSE) +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon"))
  #stat_ellipse(type = "norm", linetype = 2) +
  #stat_ellipse(type = "t")
  

alpha <- ampvis.data %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank"), rarefy = 5000) %>%
  amp_alpha_diversity(measure = c("observed", "shannon", "simpson"), richness = TRUE) %>%
  select(Time, RPM, Soil_type, ObservedOTUs, Shannon, Simpson, Chao1, ACE) %>%
  group_by(Time, RPM, Soil_type) %>%
  pivot_longer(cols = 4:8, names_to = "index", values_to = "diversity")

plot.alpha_all <- alpha %>%
  filter(index %in% c("ObservedOTUs", "Shannon")) %>%
  ggplot(aes(x = RPM, y = diversity, fill = Time)) +
  facet_grid(
    cols = vars(Soil_type),
    rows = vars(index),
    scales = "free_y") +
  geom_boxplot(outlier.size = 0) +
  geom_point(position = position_jitterdodge(jitter.width = 0.1)) +
  labs(
    x = "Intensity [RPM]",
    y = "",
    fill = "Time [s]") +
  scale_fill_manual(values = c("royalblue4", "seagreen", "salmon")) +
  theme_bw()

(plot.ord_all + plot.alpha_all) / (plot.heat_clay + plot.heat_peat + plot.heat_sand)
  
ampvis.data %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank"), rarefy = 5000) %>%
  amp_subset_samples(Soil_type == "Sand") %>%
  amp_ordinate(filter_species = 0.1, 
               type = "CCA", 
               distmeasure = "jsd",
               transform = "none",
               species_plot = FALSE,
               sample_color_by = "Time",
               sample_shape_by = "RPM",
               constrain = "RPM",
               #species_nlabels = 3,
               #species_label_taxonomy = "Genus",
               #sample_plotly = TRUE,
               sample_colorframe = FALSE) +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon"))
  #stat_ellipse(type = "norm", linetype = 2) +
  #stat_ellipse(type = "t")

tmp <- ampvis.data %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank"), rarefy = 5000) %>%
  amp_export_long(metadata_vars = c("Time", "RPM", "Soil_type")) %>%
  .[,count := as.numeric(count > 1)] %>%
  .[,keep := uniqueN(SeqID) > 1, by = .(Time, RPM, Soil_type)] %>%
  .[!is.na(OTU)] %>%
  .[,keep := NULL]

# Calculate average bray-curtis distance within replicates. 
dist.JSD <- function(aa, pseudocount=0.000001, ...) {
  KLD <- function(x,y) sum(x *log(x/y))
  JSD<- function(x,y) sqrt(0.5 * KLD(x, (x+y)/2) + 0.5 * KLD(y, (x+y)/2))
  matrixColSize <- length(colnames(aa))
  matrixRowSize <- length(rownames(aa))
  colnames <- colnames(aa)
  resultsMatrix <- matrix(0, matrixColSize, matrixColSize)
  
  inMatrix = apply(aa,1:2,function(x) ifelse (x==0,pseudocount,x))
  
  for(i in 1:matrixColSize) {
    for(j in 1:matrixColSize) { 
      resultsMatrix[i,j]=JSD(as.vector(inMatrix[,i]),
                             as.vector(inMatrix[,j]))
    }
  }
  colnames -> colnames(resultsMatrix) -> rownames(resultsMatrix)
  as.dist(resultsMatrix)->resultsMatrix
  attr(resultsMatrix, "method") <- "dist"
  return(resultsMatrix) 
}

jsd = dist.JSD(ampvis.data$abund)



tmp2 <- tmp[,.(AvgDist = {
    data.table(x = OTU, y = SeqID, z = count) %>%
    dcast(x ~ y,value.var = "z") %>%
    column_to_rownames("x") %>%
    t() %>%
    dist.JSD() %>% 
    as.vector() %>% 
    mean()
  }),by = .(Time, RPM, Soil_type)]

plot_avg_dist_1 <- tmp2 %>% 
  ggplot(aes(x = RPM, y = AvgDist, color = Time, group = Time)) +
  geom_point() +
  geom_line() +
  facet_wrap(facets = vars(Soil_type), ncol = 3) +
  labs(
    title = "The effect of beadbeating on average bray-curtis distance",
    subtitle = "Plotted as point and lines",
    y = "Average Bray-Curtis distance",
    x = "Intensity [RPM]") + 
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon")) +
  theme_bw(base_size = 14)
```

```{r}
wh_OTU_samples <- 
  {
    data_samples$abund > 2
  } %>%
  apply(1, any) %>%
  {
    .[.]
  } %>%
  names()

de_con <- 
  wh_OTU_samples[- which(wh_OTU_samples %in% intersect)]

data_samples <-
  data_samples %>%
  amp_subset_taxa(
    de_con
  )

tab_reads <- 
  data_samples$metadata %>%
  merge(
    data.frame(
      NoReads = apply(
        data_samples$abund, 2, sum)),
    by = 0
  )

tab_reads_grouped <- 
  tab_reads %>% 
  group_by(Soil_type, RPM, Time) %>%
  summarise(
    mean_reads = mean(NoReads/1000),
    sd_reads = sd(NoReads/1000)
  ) %>%
  ungroup()

# plot of reads
plot_organic <-
  tab_reads_grouped %>%
  filter(Soil_type == "Organic") %>%
  ggplot(
    aes(
      x = Time,
      y = RPM,
      fill = mean_reads
    )
  ) +
  geom_tile() +
  facet_wrap(facets = vars(Soil_type)) +
  geom_text(
    aes(
      label = paste(
        round(mean_reads, 2), "\u00B1", 
        "\n",
        signif(sd_reads, 2))
      )
    ) +
  scale_fill_gradient(
    low = "white",
    high = "seagreen",
    trans = "log10",
    limits = c(0.1, 10),
    n.breaks = 7
  ) +
  labs(
    y = "Intensity [RPM]",
    x = "Time [s]",
    fill = "Obtained reads"
  ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(legend.position = "none")

plot_clay <- 
  tab_reads_grouped %>%
  filter(Soil_type == "Clay") %>%
  ggplot(
    aes(
      x = Time,
      y = RPM,
      fill = mean_reads
    )
  ) +
  geom_tile() +
  facet_wrap(facets = vars(Soil_type)) +
  geom_text(
    aes(
      label = paste(
        round(mean_reads, 2), "\u00B1", 
        "\n",
        signif(sd_reads, 2))
      )
    ) +
  scale_fill_gradient(
    low = "white",
    high = "royalblue4",
    trans = "log10",
    limits = c(0.1, 10),
    n.breaks = 7
  ) +
  labs(
    y = "Intensity [RPM]",
    x = "Time [s]",
    fill = "Obtained reads"
  ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) +
  theme(legend.position = "none")

plot_sand <- 
  tab_reads_grouped %>%
  filter(Soil_type == "Sand") %>%
  ggplot(
    aes(
      x = Time,
      y = RPM,
      fill = mean_reads
    )
  ) +
  geom_tile() +
  facet_wrap(facets = vars(Soil_type)) +
  geom_text(
    aes(
      label = paste(
        round(mean_reads, 2), "\u00B1", 
        "\n",
        signif(sd_reads, 2))
      )
    ) +
  scale_fill_gradient(
    low = "white",
    high = "salmon",
    trans = "log10",
    limits = c(0.1, 10),
    n.breaks = 7
  ) + 
  labs(
    y = "Intensity [RPM]",
    x = "Time [s]",
    fill = "Obtained reads"
  ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) +
  theme(legend.position = "none")
```

## Rarefaction curves. 
```{r}
plot_rare_organic <-
  data_samples %>%
  amp_subset_samples(Soil_type == "Organic") %>%
  amp_rarecurve(color_by = "Time") +
  facet_grid(
    cols = vars(Soil_type),
    rows = vars(RPM)
    ) +
  labs(y = "Number of observed ASVs",
       x = "Sequencing depth [reads]",
       color = "Time [s]"
       ) +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon")) +
  theme_bw(base_size = 14) +
  theme(legend.position = c(0.75, 0.85))

plot_rare_clay <-
  data_samples %>%
  amp_subset_samples(Soil_type == "Clay") %>%
  amp_rarecurve(color_by = "Time") +
  facet_grid(
    cols = vars(Soil_type),
    rows = vars(RPM)
  ) +
  labs(y = "Number of observed ASVs",
       x = "Sequencing depth [reads]",
       color = "Time [s]"
       ) +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon")) +
  theme_bw(base_size = 14) +
  theme(legend.position = c(0.75, 0.85))

plot_rare_sand <-
  data_samples %>%
  amp_subset_samples(Soil_type == "Sand") %>%
  amp_rarecurve(color_by = "Time") +
  facet_grid(
    cols = vars(Soil_type),
    rows = vars(RPM)
  ) +
  labs(y = "Number of observed ASVs",
       x = "Sequencing depth [reads]",
       color = "Time [s]"
       ) +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon")) +
  theme_bw(base_size = 14) +
  theme(legend.position = c(0.75, 0.85))

# organize plots
fg7 <- ggplotGrob(plot_clay) %>% gtable_frame()
fg8 <- ggplotGrob(plot_organic) %>% gtable_frame()
fg9 <- ggplotGrob(plot_sand) %>% gtable_frame()
fg10 <- ggplotGrob(plot_rare_clay) %>% gtable_frame()
fg11 <- ggplotGrob(plot_rare_organic) %>% gtable_frame()
fg12 <- ggplotGrob(plot_rare_sand) %>% gtable_frame()

# combine the two 3x3 matrixes horizontally
fg_clay <-
  gtable_frame(gtable_cbind(fg7, fg10),
               width = unit(1, "null"),
               height = unit(1, "null"))

combined_2 <- 
  gtable_cbind(fg7, fg10, fg8, fg11, fg9, fg12) #combine plots

grid.newpage()

grid.draw(combined_2) #plot plots

# save plots as png.
ggsave(
  "Samples_reads_rare.png", 
  plot = combined_2, 
  width = 10, 
  height = 7)
```


```{r}
# Subset to samples with >5000 reads.
data_samples_sub <- 
  amp_subset_samples(
    data_samples, 
    minreads = 4000)

data_organic <- 
  amp_subset_samples(
    data_samples_sub, 
    Soil_type == "Organic")

data_sand <- 
  amp_subset_samples(
    data_samples_sub, 
    Soil_type == "Sand")

data_clay <- 
  amp_subset_samples(
    data_samples_sub, 
    Soil_type == "Clay")
```

We set a cutoff of at 4000 reads, samples with less than that are filtered out. Doing this we go from `r nrow(data_samples$metadata)` to `r nrow(data_samples_sub$metadata)` samples.

## $\alpha$-diversity

We asses $\alpha$-diversity after rarefying to 4000 reads pr sample. We look at the Shannon-index and number of ASVs.

```{r}
div_multcomp <- 
  amp_alphadiv(
    data_samples_sub,
  measure = c("observed", "shannon", "simpson"),
  richness = T
  ) %>%
  select(
    SeqID,
    Time,
    RPM,
    Soil_type,
    ObservedOTUs,
    Shannon,
    Simpson,
    Chao1,
    ACE
  ) %>%
  gather(
    key = index,
    value = diversity,
    -SeqID,
    -RPM,
    -Time,
    -Soil_type
  )

# alpha diversity depicted with number of OTU's and Shannon index
plot_alpha <- 
  div_multcomp %>%
  filter(index %in% c("ObservedOTUs", "Shannon")) %>%
  ggplot(
    aes(
      x = RPM,
      y = diversity,
      fill = Time
      )
    ) +
  facet_grid(
    cols = vars(Soil_type),
    rows = vars(index),
    scales = "free_y"
  ) +
  geom_boxplot(outlier.size = 0) +
  geom_point(position = position_jitterdodge(jitter.width = 0.1)) +
  labs(
    x = "Intensity [RPM]",
    y = "",
    fill = "Time [s]"
  ) +
  scale_x_discrete(limits = rev) +
  scale_fill_manual(values = c("royalblue4", "seagreen", "salmon")) +
  theme_bw()

data_obs <- 
  filter(div_multcomp, 
         index %in% "ObservedOTUs")

data_obs_organic <- 
  data_obs %>% 
  filter(Soil_type == "Organic")

data_obs_sand <- 
  data_obs %>% 
  filter(Soil_type == "Sand")

data_obs_clay <- 
  data_obs %>% 
  filter(Soil_type == "Clay")

# overall
aov_all <- 
  anova(lm(diversity ~ Time * RPM + Soil_type, data = data_obs)) %>%
  data.frame() %>%
  rownames_to_column("Variable") %>%
  mutate(R2 = Sum.Sq / sum(Sum.Sq) * 100) %>%
  mutate(pval = ifelse(Pr..F. < 0.001, "<0.001", round(Pr..F., 2))) %>%
  select(Variable, R2, pval)

# organic
aov_organic <- 
  anova(lm(diversity ~ Time * RPM, data = data_obs_organic)) %>%
  data.frame() %>%
  rownames_to_column("Variable") %>%
  mutate(R2 = Sum.Sq / sum(Sum.Sq) * 100) %>%
  mutate(pval = ifelse(Pr..F. < 0.001, "<0.001", round(Pr..F., 2))) %>%
  select(Variable, R2, pval)

# sand
aov_sand <- 
  anova(lm(diversity ~ Time * RPM, data = data_obs_sand)) %>%
  data.frame() %>%
  rownames_to_column("Variable") %>%
  mutate(R2 = Sum.Sq / sum(Sum.Sq) * 100) %>%
  mutate(pval = ifelse(Pr..F. < 0.001, "<0.001", round(Pr..F., 2))) %>%
  select(Variable, R2, pval)

# clay
aov_clay <- 
  anova(lm(diversity ~ Time * RPM, data = data_obs_clay)) %>%
  data.frame() %>%
  rownames_to_column("Variable") %>%
  mutate(R2 = Sum.Sq / sum(Sum.Sq) * 100) %>%
  mutate(pval = ifelse(Pr..F. < 0.001, "<0.001", round(Pr..F., 2))) %>%
  select(Variable, R2, pval)
```

## $\beta$-diversity
First an overview by calculating bray-curtis distances between samples, based on present/absent data. A clear separation between soil_types which make sense!
```{r}
# pca-plot
plot_pca_all <-   
  data_samples_sub %>%
  amp_ordinate(
    filter_species = 0,
    distmeasure = "bray",
    transform = "pa",
    sample_color_by = "Time",
    sample_shape_by = "RPM",
    sample_trajectory_group = "Soil_type",
    sample_colorframe = F,
    sample_colorframe_label = "Soil_type",
    sample_colorframe_label_size = 6,
    print_caption = T
    ) +
  labs(
    title = "PCA-plot for all samples"
    ) +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon")) +
  xlim(-7.5, 7.5) +
  ylim(-10, 10) +
  theme_bw(base_size = 14) +
  guides(color = F)

plot_pca_clay <-   
  data_clay %>%
  amp_ordinate(
    filter_species = 0,
    distmeasure = "bray",
    transform = "pa",
    sample_color_by = "Time",
    sample_shape_by = "RPM",
    sample_trajectory_group = "Soil_type",
    sample_colorframe = T,
    print_caption = T
    ) +
  labs(
    title = "PCA-plot for clay samples"
    ) +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon")) +
  scale_fill_manual(values = c("royalblue4", "seagreen", "salmon")) +
  xlim(-7.5, 7.5) +
  ylim(-10, 10) +
  theme_bw(base_size = 14)

plot_pca_organic <-   
  data_organic %>%
  amp_ordinate(
    filter_species = 0,
    distmeasure = "bray",
    transform = "pa",
    sample_color_by = "Time",
    sample_shape_by = "RPM",
    sample_trajectory_group = "Soil_type",
    sample_colorframe = T,
    print_caption = T
    ) +
  labs(
    title = "PCA-plot for organic samples"
    ) +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon")) +
  scale_fill_manual(values = c("royalblue4", "seagreen", "salmon")) +
  xlim(-7.5, 7.5) +
  ylim(-10, 10) +
  theme_bw(base_size = 14)

plot_pca_sand <-   
  data_sand %>%
  amp_ordinate(
    filter_species = 0,
    distmeasure = "bray",
    transform = "pa",
    sample_color_by = "Time",
    sample_shape_by = "RPM",
    sample_trajectory_group = "Soil_type",
    sample_colorframe = T,
    print_caption = T
    ) +
  labs(
    title = "PCA-plot for sand samples"
    ) +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon")) +
  scale_fill_manual(values = c("royalblue4", "seagreen", "salmon")) +
  xlim(-7.5, 7.5) +
  ylim(-10, 10) +
  theme_bw(base_size = 14)

combined_3 <- 
  plot_pca_all + 
  plot_pca_clay + 
  plot_pca_organic + 
  plot_pca_sand & 
  theme(legend.position = "bottom")

combined_3 + 
  plot_layout(
    guides = "collect"
    )
```

```{r}
tmp <- amp_export_long(
  data_samples_sub,
  metadata_vars = c("Time", "RPM", "Soil_type")) %>%
  .[,count := as.numeric(count > 1)] %>%
  .[,keep := uniqueN(SeqID) > 1, by = .(Time, RPM, Soil_type)] %>%
  .[!is.na(OTU)] %>%
  .[,keep := NULL]

# Calculate average bray-curtis distance within replicates. 
tmp2 <- tmp[,.(AvgDist = {
    data.table(x = OTU, y = SeqID, z = count) %>%
    dcast(x ~ y,value.var = "z") %>%
    column_to_rownames("x") %>%
    t() %>%
    vegdist(method = "bray") %>% 
    as.vector() %>% 
    mean()
  }),by = .(Time, RPM, Soil_type)]

plot_avg_dist_1 <-
  tmp2 %>%
  ggplot(
    aes(
      x = RPM,
      y = AvgDist,
      color = Time,
      group = Time
      )
    ) +
    geom_point() +
    geom_line() +
    facet_wrap(facets = vars(Soil_type), ncol = 3) +
    labs(x = "", y = "Average distance within replicates") +
    labs(
      title = "The effect of beadbeating on average bray-curtis distance",
      subtitle = "Plotted as point and lines",
      y = "Average Bray-Curtis distance",
      x = "Intensity [RPM]"
      ) +
    scale_color_manual(values = c("royalblue4", "seagreen", "salmon")) +
    theme_bw(base_size = 14)

plot_avg_dist_clay <-
  tmp2 %>%
  filter(
    Soil_type == "Clay"
    ) %>%
  ggplot(
    aes(
      x = Time,
      y = RPM,
      fill = AvgDist
      )
    ) +
  geom_tile() +
  facet_grid(cols = vars(Soil_type)) +
  geom_text(aes(label = paste(round(AvgDist, 4)))) +
  scale_fill_gradient(
    low = "royalblue4",
    high = "white",
    trans = "log10"
  ) +
  labs(
    title = "The effect of beadbeating on average bray-curtis distance",
    subtitle = "Plotted as heatmap",
    y = "Intensity [RPM]",
    x = "Duration [s]",
    fill = "Avg. distance"
    ) +
  theme_bw(base_size = 14)
  
plot_avg_dist_organic <-
  tmp2 %>%
  filter(
    Soil_type == "Organic"
    ) %>%
  ggplot(
    aes(
      x = Time,
      y = RPM,
      fill = AvgDist
      )
    ) +
  geom_tile() +
  facet_grid(cols = vars(Soil_type)) +
  geom_text(aes(label = paste(round(AvgDist, 4)))) +
  scale_fill_gradient(
    low = "seagreen",
    high = "white",
    trans = "log10"
  ) +
  labs(
    title = "The effect of beadbeating on average bray-curtis distance",
    subtitle = "Plotted as heatmap",
    y = "Intensity [RPM]",
    x = "Duration [s]",
    fill = "Avg. distance"
    ) +
  theme_bw(base_size = 14)
  
plot_avg_dist_sand <-
  tmp2 %>%
  filter(
    Soil_type == "Sand"
    ) %>%
  ggplot(
    aes(
      x = Time,
      y = RPM,
      fill = AvgDist
      )
    ) +
  geom_tile() +
  facet_grid(cols = vars(Soil_type)) +
  geom_text(aes(label = paste(round(AvgDist, 4)))) +
  scale_fill_gradient(
    low = "salmon",
    high = "white",
    trans = "log10"
  ) +
  labs(
    title = "The effect of beadbeating on average bray-curtis distance",
    subtitle = "Plotted as heatmap",
    y = "Intensity [RPM]",
    x = "Duration [s]",
    fill = "Avg. distance"
    ) +
  theme_bw(base_size = 14)

combined_4 <- 
  plot_avg_dist_1 +
  plot_avg_dist_clay +
  plot_avg_dist_organic +
  plot_avg_dist_sand

combined_4
```

*Both time and RPM seems to lower the average bray-curtis distance within replicates. The trend is however different for the different types of soil. *

## Heatmaps
```{r}
heat_clay_1 <-
  data_clay %>%
  amp_subset_samples(
    Time == "120"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "royalblue4"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("Gemmatimonadetes",
                   "Firmicutes",
                   "Rokubacteria",
                   "Thaumarchaeota",
                   "Planctomycetes",
                   "Chloroflexi",
                   "Verrucomicrobia",
                   "Acidobacteria",
                   "Actinobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "Heatmap for clay samples",
    subtitle = "Time = 120 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none")
  
heat_clay_2 <-
  data_clay %>%
  amp_subset_samples(
    Time == "240"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "royalblue4"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("Gemmatimonadetes",
                   "Firmicutes",
                   "Rokubacteria",
                   "Thaumarchaeota",
                   "Planctomycetes",
                   "Chloroflexi",
                   "Verrucomicrobia",
                   "Acidobacteria",
                   "Actinobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "",
    subtitle = "Time = 240 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank())
  
heat_clay_3 <-
  data_clay %>%
  amp_subset_samples(
    Time == "360"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "royalblue4"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("Gemmatimonadetes",
                   "Firmicutes",
                   "Rokubacteria",
                   "Thaumarchaeota",
                   "Planctomycetes",
                   "Chloroflexi",
                   "Verrucomicrobia",
                   "Acidobacteria",
                   "Actinobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "",
    subtitle = "Time = 360 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none",
        axis.text.y = element_blank())

combine_clay <- 
  heat_clay_1 +
  heat_clay_2 +
  heat_clay_3 

combine_clay + 
  plot_layout(
    guides = "collect"
    )

heat_organic_1 <-
  data_organic %>%
  amp_subset_samples(
    Time == "120"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "seagreen"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("WPS-2",
                   "Bacteroidetes",
                   "Thaumarchaeota",
                   "Firmicutes",
                   "Chloroflexi",
                   "Planctomycetes",
                   "Actinobacteria",
                   "Verrucomicrobia",
                   "Proteobacteria",
                   "Acidobacteria")
    ) +
  labs(
    title = "Heatmap for organic samples",
    subtitle = "Time = 120 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none")
  
heat_organic_2 <-
  data_organic %>%
  amp_subset_samples(
    Time == "240"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "seagreen"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("WPS-2",
                   "Bacteroidetes",
                   "Thaumarchaeota",
                   "Firmicutes",
                   "Chloroflexi",
                   "Planctomycetes",
                   "Actinobacteria",
                   "Verrucomicrobia",
                   "Proteobacteria",
                   "Acidobacteria")
    ) +
  labs(
    title = "",
    subtitle = "Time = 240 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank())
  
heat_organic_3 <-
  data_organic %>%
  amp_subset_samples(
    Time == "360"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "seagreen"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("WPS-2",
                   "Bacteroidetes",
                   "Thaumarchaeota",
                   "Firmicutes",
                   "Chloroflexi",
                   "Planctomycetes",
                   "Actinobacteria",
                   "Verrucomicrobia",
                   "Proteobacteria",
                   "Acidobacteria")
    ) +
  labs(
    title = "",
    subtitle = "Time = 360 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none",
        axis.text.y = element_blank())
  
combine_organic <- 
  heat_organic_1 +
  heat_organic_2 +
  heat_organic_3 

combine_organic + 
  plot_layout(
    guides = "collect"
    )

heat_sand_1 <-
  data_sand %>%
  amp_subset_samples(
    Time == "120"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "salmon"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("GAL15",
                   "Firmicutes",
                   "Gemmatimonadetes",
                   "WPS-2",
                   "Verrucomicrobia",
                   "Planctomycetes",
                   "Thaumarchaeota",
                   "Actinobacteria",
                   "Acidobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "Heatmap for sand samples",
    subtitle = "Time = 120 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none")
  
heat_sand_2 <-
  data_sand %>%
  amp_subset_samples(
    Time == "240"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "salmon"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("GAL15",
                   "Firmicutes",
                   "Gemmatimonadetes",
                   "WPS-2",
                   "Verrucomicrobia",
                   "Planctomycetes",
                   "Thaumarchaeota",
                   "Actinobacteria",
                   "Acidobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "",
    subtitle = "Time = 240 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank())
  
heat_sand_3 <-
  data_sand %>%
  amp_subset_samples(
    Time == "360"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "salmon"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("GAL15",
                   "Firmicutes",
                   "Gemmatimonadetes",
                   "WPS-2",
                   "Verrucomicrobia",
                   "Planctomycetes",
                   "Thaumarchaeota",
                   "Actinobacteria",
                   "Acidobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "",
    subtitle = "Time = 360 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none",
        axis.text.y = element_blank())
  
combine_sand <- 
  heat_sand_1 +
  heat_sand_2 +
  heat_sand_3 

combine_sand + 
  plot_layout(
    guides = "collect"
    )
```

## Faceted better
```{r}
heat_clay_1.1 <-
  data_clay %>%
  amp_subset_samples(
    RPM == "1200"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "royalblue4"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("Gemmatimonadetes",
                   "Firmicutes",
                   "Rokubacteria",
                   "Thaumarchaeota",
                   "Planctomycetes",
                   "Chloroflexi",
                   "Verrucomicrobia",
                   "Acidobacteria",
                   "Actinobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "Heatmap for clay samples",
    subtitle = "RPM = 1200"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none")
  
heat_clay_1.2 <-
  data_clay %>%
  amp_subset_samples(
    RPM == "1400"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "royalblue4"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("Gemmatimonadetes",
                   "Firmicutes",
                   "Rokubacteria",
                   "Thaumarchaeota",
                   "Planctomycetes",
                   "Chloroflexi",
                   "Verrucomicrobia",
                   "Acidobacteria",
                   "Actinobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "",
    subtitle = "RPM = 1400 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank(),
        legend.position = "none")
  
heat_clay_1.3 <-
  data_clay %>%
  amp_subset_samples(
    RPM == "1600"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "royalblue4"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("Gemmatimonadetes",
                   "Firmicutes",
                   "Rokubacteria",
                   "Thaumarchaeota",
                   "Planctomycetes",
                   "Chloroflexi",
                   "Verrucomicrobia",
                   "Acidobacteria",
                   "Actinobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "",
    subtitle = "RPM = 1600 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none",
        axis.text.y = element_blank())

heat_clay_1.4 <-
  data_clay %>%
  amp_subset_samples(
    RPM == "1800"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "royalblue4"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("Gemmatimonadetes",
                   "Firmicutes",
                   "Rokubacteria",
                   "Thaumarchaeota",
                   "Planctomycetes",
                   "Chloroflexi",
                   "Verrucomicrobia",
                   "Acidobacteria",
                   "Actinobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "",
    subtitle = "RPM = 1800 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank())

combine_clay <- 
  heat_clay_1.1 +
  heat_clay_1.2 +
  heat_clay_1.3 +
  heat_clay_1.4

plot_clay <- 
  combine_clay + 
  plot_layout(
    guides = "collect",
    ncol = 4
    )
```

```{r}
heat_organic_1.1 <-
  data_organic %>%
  amp_subset_samples(
    RPM == "1200"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "seagreen"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("WPS-2",
                   "Bacteroidetes",
                   "Thaumarchaeota",
                   "Firmicutes",
                   "Chloroflexi",
                   "Planctomycetes",
                   "Actinobacteria",
                   "Verrucomicrobia",
                   "Proteobacteria",
                   "Acidobacteria")
    ) +
  labs(
    title = "Heatmap for organic samples",
    subtitle = "RPM = 1200"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none")
  
heat_organic_1.2 <-
  data_organic %>%
  amp_subset_samples(
    RPM == "1400"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "seagreen"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("WPS-2",
                   "Bacteroidetes",
                   "Thaumarchaeota",
                   "Firmicutes",
                   "Chloroflexi",
                   "Planctomycetes",
                   "Actinobacteria",
                   "Verrucomicrobia",
                   "Proteobacteria",
                   "Acidobacteria")
    ) +
  labs(
    title = "",
    subtitle = "RPM = 1400 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank(),
        legend.position = "none")
  
heat_organic_1.3 <-
  data_organic %>%
  amp_subset_samples(
    RPM == "1600"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "seagreen"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("WPS-2",
                   "Bacteroidetes",
                   "Thaumarchaeota",
                   "Firmicutes",
                   "Chloroflexi",
                   "Planctomycetes",
                   "Actinobacteria",
                   "Verrucomicrobia",
                   "Proteobacteria",
                   "Acidobacteria")
    ) +
  labs(
    title = "",
    subtitle = "RPM = 1600 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none",
        axis.text.y = element_blank())

heat_organic_1.4 <-
  data_organic %>%
  amp_subset_samples(
    RPM == "1800"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "seagreen"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("WPS-2",
                   "Bacteroidetes",
                   "Thaumarchaeota",
                   "Firmicutes",
                   "Chloroflexi",
                   "Planctomycetes",
                   "Actinobacteria",
                   "Verrucomicrobia",
                   "Proteobacteria",
                   "Acidobacteria")
    ) +
  labs(
    title = "",
    subtitle = "RPM = 1800 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank())

combine_organic <- 
  heat_organic_1.1 +
  heat_organic_1.2 +
  heat_organic_1.3 +
  heat_organic_1.4

plot_organic <- 
  combine_organic + 
  plot_layout(
    guides = "collect",
    ncol = 4
    )
```

```{r}
heat_sand_1.1 <-
  data_sand %>%
  amp_subset_samples(
    RPM == "1200"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "salmon"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("GAL15",
                   "Firmicutes",
                   "Gemmatimonadetes",
                   "WPS-2",
                   "Verrucomicrobia",
                   "Planctomycetes",
                   "Thaumarchaeota",
                   "Actinobacteria",
                   "Acidobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "Heatmap for sand samples",
    subtitle = "RPM = 1200"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none")
  
heat_sand_1.2 <-
  data_sand %>%
  amp_subset_samples(
    RPM == "1400"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "salmon"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("GAL15",
                   "Firmicutes",
                   "Gemmatimonadetes",
                   "WPS-2",
                   "Verrucomicrobia",
                   "Planctomycetes",
                   "Thaumarchaeota",
                   "Actinobacteria",
                   "Acidobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "",
    subtitle = "RPM = 1400 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank(),
        legend.position = "none")
  
heat_sand_1.3 <-
  data_sand %>%
  amp_subset_samples(
    RPM == "1600"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "salmon"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("GAL15",
                   "Firmicutes",
                   "Gemmatimonadetes",
                   "WPS-2",
                   "Verrucomicrobia",
                   "Planctomycetes",
                   "Thaumarchaeota",
                   "Actinobacteria",
                   "Acidobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "",
    subtitle = "RPM = 1600 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none",
        axis.text.y = element_blank())

heat_sand_1.4 <-
  data_sand %>%
  amp_subset_samples(
    RPM == "1800"
    ) %>%
  amp_heatmap(
    group_by = "SeqID",
    facet_by = "Time",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "salmon"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("GAL15",
                   "Firmicutes",
                   "Gemmatimonadetes",
                   "WPS-2",
                   "Verrucomicrobia",
                   "Planctomycetes",
                   "Thaumarchaeota",
                   "Actinobacteria",
                   "Acidobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "",
    subtitle = "RPM = 1800 s"
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank())

combine_sand <- 
  heat_sand_1.1 +
  heat_sand_1.2 +
  heat_sand_1.3 +
  heat_sand_1.4

plot_sand <- 
  combine_sand + 
  plot_layout(
    guides = "collect",
    ncol = 4
    )
```

## Combined 
```{r}
heat_clay_2.1 <-
  data_clay %>%
  amp_heatmap(
    group_by = "Time",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "royalblue4"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("Gemmatimonadetes",
                   "Firmicutes",
                   "Rokubacteria",
                   "Thaumarchaeota",
                   "Planctomycetes",
                   "Chloroflexi",
                   "Verrucomicrobia",
                   "Acidobacteria",
                   "Actinobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "Heatmap for clay samples",
    subtitle = "..."
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none")
  
heat_organic_2.1 <-
  data_organic %>%
  amp_heatmap(
    group_by = "Time",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "seagreen"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("WPS-2",
                   "Bacteroidetes",
                   "Thaumarchaeota",
                   "Firmicutes",
                   "Chloroflexi",
                   "Planctomycetes",
                   "Actinobacteria",
                   "Verrucomicrobia",
                   "Proteobacteria",
                   "Acidobacteria")
    ) +
  labs(
    title = "Heatmap for organic samples",
    subtitle = "..."
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none")

heat_sand_2.1 <-
  data_sand %>%
  amp_heatmap(
    group_by = "Time",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "salmon"),
    plot_colorscale = "log10",
    plot_values = TRUE,
    order_y_by = c("GAL15",
                   "Firmicutes",
                   "Gemmatimonadetes",
                   "WPS-2",
                   "Verrucomicrobia",
                   "Planctomycetes",
                   "Thaumarchaeota",
                   "Actinobacteria",
                   "Acidobacteria",
                   "Proteobacteria")
    ) +
  labs(
    title = "Heatmap for sand samples",
    subtitle = "..."
    ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        legend.position = "none")

combine_average <- 
  heat_clay_2.1 +
  heat_organic_2.1 +
  heat_sand_2.1

plot_average <- 
  combine_average + 
  plot_layout(
    guides = "collect",
    ncol = 3
    )

# combine heatmaps plus averages
```

## Differential analysis

```{r}
# Diversity mean
wh_OTU_high <-
  {
    data_samples_sub$abund > 20
  } %>%
  apply(1, any) %>%
  {
    .[.]
  } %>%
  names()

data_samples_sub_high <- data_samples_sub %>%
  amp_subset_taxa(tax_vector = wh_OTU_high)

data_samples_sub_high2 <- data_samples_sub_high %>%
  amp_subset_taxa(tax_vector = wh_OTU_PCR, remove = T)

plotx <- data_samples_sub_high %>%
  amp_subset_samples(Soil_type == "Clay") %>%
  amp_heatmap(
    group_by = "LibID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "burlywood4"),
    plot_colorscale = "log10",
    plot_values = TRUE
  ) +
  ggtitle("Heatmap for Clay samples no OTUs removed",
    subtitle = "Triplicates faceted by bead beating intensity and duration"
  )

ploty <- data_samples_sub_high2 %>%
  amp_subset_samples(Soil_type == "Clay") %>%
  amp_heatmap(
    group_by = "LibID",
    facet_by = "RPM",
    tax_aggregate = "Phylum",
    tax_show = 10,
    color_vector = c("white", "burlywood4"),
    plot_colorscale = "log10",
    plot_values = TRUE
  ) +
  ggtitle("Heatmap for Clay samples PCR OTUs removed",
    subtitle = "Triplicates faceted by bead beating intensity and duration"
  )

grid.arrange(plotx, ploty, nrow = 2)

data_samples_sub_high_long <- data_samples_sub_high %>%
  amp_export_long(
    metadata_vars = c("SeqID", "Soil_type", "RPM", "Time"),
    tax_levels = "OTU"
  )

tmp_organic_high <- data_samples_sub_high_long %>%
  group_by(Soil_type) %>%
  filter(count > 0) %>%
  distinct(OTU) %>%
  filter(Soil_type == "Organic") %>%
  nrow()

tmp_sand_high <- data_samples_sub_high_long %>%
  group_by(Soil_type) %>%
  filter(count > 0) %>%
  distinct(OTU) %>%
  filter(Soil_type == "Sand") %>%
  nrow()

tmp_clay_high <- data_samples_sub_high_long %>%
  group_by(Soil_type) %>%
  filter(count > 0) %>%
  distinct(OTU) %>%
  filter(Soil_type == "Clay") %>%
  nrow()

tmp_high <- data_samples_sub_high_long %>%
  filter(count > 0) %>%
  group_by(SeqID, Time, RPM, Soil_type) %>%
  summarise(diversity = n()) %>%
  ungroup() %>%
  group_by(Time, RPM, Soil_type) %>%
  mutate(diversity.mean = mean(diversity)) %>%
  ungroup() %>%
  group_by(Soil_type) %>%
  mutate(diversity.pct = case_when(
    Soil_type == "Organic" ~ 100 * diversity.mean / tmp_organic_high,
    Soil_type == "Sand" ~ 100 * diversity.mean / tmp_sand_high,
    Soil_type == "Clay" ~ 100 * diversity.mean / tmp_clay_high
  ))

plota <- ggplot(
  tmp_high,
  aes(x = Time, y = RPM, fill = diversity.pct)
) +
  geom_tile() +
  facet_grid(cols = vars(Soil_type)) +
  geom_text(aes(label = round(diversity.pct, 1))) +
  scale_fill_gradient(low = "white", high = "darkorange") +
  ggtitle("Number of different ASVs, n > 20") +
  xlab("Duration [s]") +
  ylab("Intensity [RPM]") +
  theme_bw()

wh_OTU_all <-
  {
    data_samples_sub$abund < 20 & data_samples_sub$abund > 5
  } %>%
  apply(1, any) %>%
  {
    .[.]
  } %>%
  names()

data_samples_sub_all <- amp_subset_taxa(data_samples_sub, tax_vector = wh_OTU_all)

data_samples_sub_all2 <- amp_subset_taxa(data_samples_sub, tax_vector = wh_OTU_all) %>%
  amp_subset_taxa(tax_vector = wh_OTU_PCR, remove = T)

data_samples_sub_all_long <- data_samples_sub_all %>% amp_export_long(metadata_vars = c("SeqID", "Soil_type", "RPM", "Time"), tax_levels = "OTU")

tmp_organic_all <- data_samples_sub_all_long %>%
  group_by(Soil_type) %>%
  filter(count > 0) %>%
  distinct(OTU) %>%
  filter(Soil_type == "Organic") %>%
  nrow()

tmp_sand_all <- data_samples_sub_all_long %>%
  group_by(Soil_type) %>%
  filter(count > 0) %>%
  distinct(OTU) %>%
  filter(Soil_type == "Sand") %>%
  nrow()

tmp_clay_all <- data_samples_sub_all_long %>%
  group_by(Soil_type) %>%
  filter(count > 0) %>%
  distinct(OTU) %>%
  filter(Soil_type == "Clay") %>%
  nrow()

tmp_all <- data_samples_sub_all_long %>%
  filter(count > 0) %>%
  group_by(SeqID, Time, RPM, Soil_type) %>%
  summarise(diversity = n()) %>%
  ungroup() %>%
  group_by(Time, RPM, Soil_type) %>%
  summarise(diversity.mean = mean(diversity)) %>%
  ungroup() %>%
  group_by(Soil_type) %>%
  mutate(diversity.pct = case_when(
    Soil_type == "Organic" ~ 100 * diversity.mean / tmp_organic_all,
    Soil_type == "Sand" ~ 100 * diversity.mean / tmp_sand_all,
    Soil_type == "Clay" ~ 100 * diversity.mean / tmp_clay_all
  ))

plotb <- ggplot(
  tmp_all,
  aes(x = Time, y = RPM, fill = diversity.pct)
) +
  geom_tile() +
  facet_grid(cols = vars(Soil_type)) +
  geom_text(aes(label = round(diversity.pct, 1))) +
  scale_fill_gradient(low = "white", high = "darkorange") +
  ggtitle("Number of different ASVs, 20 > n > 5") +
  xlab("Duration [s]") +
  ylab("Intensity [RPM]") +
  theme_bw()

grid.arrange(plot3, plot4, plota, plotb, ncol = 2)
```

*The diversity percent generally seem to fall for highly abundant zOTUs and increase for less abundant ones with both RPM and Time.*

```{r}
# MA-analysis
data_bead_opt_organic <- data_organic %>% amp_subset_samples(RPM %in% c("1600", "1800") & Time == "360")

data_bead_opt_sand <- data_sand %>% amp_subset_samples(RPM %in% c("1600", "1800") & Time == "360")

data_bead_opt_clay <- data_clay %>% amp_subset_samples(RPM %in% c("1600", "1800") & Time == "360")

deseq_organic <- data_bead_opt_organic %>% amp_diffabund(group = "RPM", fold = 1)

deseq_sand <- data_bead_opt_organic %>% amp_diffabund(group = "RPM", fold = 1)

deseq_clay <- data_bead_opt_organic %>% amp_diffabund(group = "RPM", fold = 1)
```

*No significant difference between 1600 and 1800 RPM for any of the soil types at 360 s.*

## MA-plots - extended
We use a MA-plot to visualize the change of the ASVs of the nine most abundant phyla of bacteria and one archea between the two analysis. A baseMean read abundance cut-off is set at > 5.
```{r}
ggplot(df_res, aes(x = baseMean, y = log2FoldChange)) +
  geom_point() +
  scale_x_log10() +
  geom_vline(xintercept = 5) +
  geom_hline(yintercept = 1) +
  geom_hline(yintercept = -1) +
  geom_text(aes(label = ifelse(df_res$baseMean > 5 & abs(df_res$log2FoldChange) > 1, as.character(Tax), "")), hjust = 1, vjust = 1) +
  # annotate(geom="text", x=100, y=4, label="benchmarking2", color="black") +
  # annotate(geom="text", x=100, y=-4, label="beadbeating", color="black") +
  labs(x = "baseMean read abundance", y = "log2 fold change") +
  theme_classic()
```

```{r}
sub_sand <- amp_subset_samples(data_sand, RPM %in% c("1600", "1800") & Time == "360")

result <- amp_diffabund(sub_sand, group = "RPM", tax_aggregate = "OTU", signif_thrh = 0.1)

df_res <- as.data.frame(result$plot_MA$data)

# df_res$OTU <- df_res$Tax %>% as.character()

# df_res <- left_join(tax, df_res, by = "OTU") %>% filter(!is.na(baseMean))

# df_res <- df_res %>% filter(Phylum != "") %>% filter(Phylum %in% c("p__Proteobacteria", "p__Synergistetes", "p__Actinobacteria", "p__Cloacimonetes", "p__Euryarchaeota", "p__Firmicutes", "p__Chloroflexi", "p__Bacteroidetes"))

# df_res$Phylum <- factor(df_res$Phylum)

ggplot(df_res, aes(x = baseMean, y = log2FoldChange)) +
  geom_point() +
  scale_x_log10() +
  geom_vline(xintercept = 5) +
  geom_hline(yintercept = 1) +
  geom_hline(yintercept = -1) +
  geom_text(aes(label = ifelse(df_res$baseMean > 5 & abs(df_res$log2FoldChange) > 1, as.character(Tax), "")), hjust = 1, vjust = 1) +
  # annotate(geom="text", x=100, y=4, label="benchmarking2", color="black") +
  # annotate(geom="text", x=100, y=-4, label="beadbeating", color="black") +
  labs(x = "baseMean read abundance", y = "log2 fold change") +
  theme_classic()
```

```{r}
sub_clay <- amp_subset_samples(data_clay, RPM %in% c("1600", "1800") & Time == "360")

result <- amp_diffabund(sub_clay, group = "RPM", tax_aggregate = "OTU", signif_thrh = 1)

df_res <- as.data.frame(result$plot_MA$data)

# df_res$OTU <- df_res$Tax %>% as.character()

# df_res <- left_join(tax, df_res, by = "OTU") %>% filter(!is.na(baseMean))

# df_res <- df_res %>% filter(Phylum != "") %>% filter(Phylum %in% c("p__Proteobacteria", "p__Synergistetes", "p__Actinobacteria", "p__Cloacimonetes", "p__Euryarchaeota", "p__Firmicutes", "p__Chloroflexi", "p__Bacteroidetes"))

# df_res$Phylum <- factor(df_res$Phylum)

ggplot(df_res, aes(x = baseMean, y = log2FoldChange)) +
  geom_point() +
  scale_x_log10() +
  geom_vline(xintercept = 5) +
  geom_hline(yintercept = 1) +
  geom_hline(yintercept = -1) +
  geom_text(aes(label = ifelse(df_res$baseMean > 5 & abs(df_res$log2FoldChange) > 1, as.character(Tax), "")), hjust = 1, vjust = 1) +
  # annotate(geom="text", x=100, y=4, label="benchmarking2", color="black") +
  # annotate(geom="text", x=100, y=-4, label="beadbeating", color="black") +
  labs(x = "baseMean read abundance", y = "log2 fold change") +
  theme_classic()
```




```{r}
datraw %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank")) %>%
  amp_subset_samples(Soil_type == "Clay") %>%
  amp_ordinate(filter_species = 0.1, 
               type = "RDA",
               distmeasure = "bray",
               transform = "none",
               species_plot = FALSE,
               sample_color_by = "Time",
               sample_shape_by = "RPM",
               constrain = "RPM",
               #species_nlabels = 3,
               #species_label_taxonomy = "Genus",
               #sample_plotly = TRUE,
               sample_colorframe = FALSE) +
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon"))
  #stat_ellipse(type = "norm", linetype = 2) +
  #stat_ellipse(type = "t")

tmp <- datraw %>%
  amp_subset_samples(!Soil_type %in% c("PCR-POS", "PCR-NEG", "Blank"), rarefy = 5000) %>%
  amp_export_long(metadata_vars = c("Time", "RPM", "Soil_type", "Beadbeating")) %>%
  .[,count := as.numeric(count > 1)] %>%
  .[,keep := uniqueN(SeqID) > 1, by = .(Time, RPM, Soil_type, Beadbeating)] %>%
  .[!is.na(OTU)] %>%
  .[,keep := NULL]

# Calculate average bray-curtis distance within replicates. 

tmp2 <- tmp[,.(AvgDist = {
    data.table(x = OTU, y = SeqID, z = count) %>%
    dcast(x ~ y,value.var = "z") %>%
    column_to_rownames("x") %>%
    t() %>%
    vegan::vegdist(method = "bray") %>% 
    as.vector() %>% 
    mean()
  }),by = .(Time, RPM, Soil_type, Beadbeating)]

plot_avg_dist_1 <- tmp2 %>% 
  ggplot(aes(x = RPM, y = AvgDist, color = Time, group = Time)) +
  geom_point() +
  geom_line() +
  facet_wrap(facets = vars(Soil_type), ncol = 3) +
  labs(
    title = "The effect of beadbeating on average bray-curtis distance",
    subtitle = "Plotted as point and lines",
    y = "Average Bray-Curtis distance",
    x = "Intensity [RPM]") + 
  scale_color_manual(values = c("royalblue4", "seagreen", "salmon")) +
  theme_bw(base_size = 14)
```
