---
title: "Amplicon diversity of MFD standard samples 2.0 - QC"
author: "Thomas Bygh Nymann Jensen, Thomas Y. Michaelsen"
date: "`r format(Sys.time(), '%d-%m-%Y')`, Aalborg, Denmark"
output: 
  bookdown::html_document2:
    toc: true
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE)

library(here)
setwd(here("Benchmarking"))

# Load the needed packages.
library(data.table)
library(readxl)
library(tidyverse)
library(ampvis2)
# library(DESeq2)
library(vegan)
library(patchwork)
library(ggpubr)

### Load and tidy ###
source("loaddata.R")
colors <- c("grey", "darkgoldenrod2", "royalblue4", "seagreen", "salmon", "purple")
```

## QC - PCR controls
Before data is filtered the PCR controls of plate A, B and C are evaluated for number of reads. 
Number of reads for the positive and negative controls. Same positive control used. Concentrations are alike.

```{r}
dat_PCR <- amp_subset_samples(datraw,is.na(Kit))

tabReads <- merge(
  dat_PCR$metadata,
  data.frame(NoReads = apply(dat_PCR$abund,2,sum)),
  by = 0)

ggplot(tabReads,aes(x = LibID,y = NoReads,colour = LibID)) + 
  geom_jitter(width = .1) +
  labs(colour = "PCR controls",x = "",y = "# of reads") +
  theme(axis.text.x = element_text(angle=90, hjust=1))
```

## QC - blanks
Number of reads for all blanks across kits. Looks OK, except for ZymoMagbead HT kit where there is some contamination that likely skews the data for this kit - possibly cross contamination with Activated sludge (from PCoA) - this plot also shows problems for MIDAS HT on a lot of samples. 
```{r}
dat_blank <- amp_subset_samples(datraw,Soil_type == "Blank")

tabReads <- merge(
  dat_blank$metadata,
  data.frame(NoReads = apply(dat_blank$abund,2,sum)),
  by = 0)

ggplot(tabReads,aes(x = Kit, y = NoReads, colour = Kit)) + 
  geom_jitter(width = .1) +
  theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(x = "",y = "# of reads") +
  theme(legend.position = "none")

datraw %>%
  amp_ordinate(sample_color_by = "Soil_type",
               type = "PCA",
               sample_shape_by = "Kit",
               transform = "hellinger")
               #sample_plotly = F)
               #species_nlabels = 10,
               #species_label_taxonomy = "Genus")
```

## QC - samples

```{r}
dat_types <- datraw %>% 
  amp_subset_samples(!Soil_type %in% c("Blank","Activated sludge") & !is.na(Kit))
```

For FastSpin, MIDAS HT and ZymoMagbead HT we observe that especially the Organic samples fails to produce successful libraries resulting in low amount of reads. This is possible related directly to PCR inhibition due to lack of removal of humic substances. 
```{r}
tabReads <- merge(
  dat_types$metadata,
  data.frame(NumReads = apply(dat_types$abund, 2, sum)),
  by = 0)

plot.Num_reads <- tabReads %>%
  ggplot(aes(x = Soil_type, y = NumReads, fill = Soil_type)) + 
  facet_wrap(facets = vars(Kit), nrow = 1) +
  geom_boxplot(alpha = 0.5) +
  geom_point(position = position_jitterdodge(jitter.width = .1)) +
  theme(legend.position = "right") +
  labs(x = "",
       y = "",
       title = "Number of reads across sample types") +
  scale_fill_manual(values = colors, name = "") +
  theme_bw(base_size = 10) +
  theme(legend.position = "none",
      axis.text.x = element_blank(),
      axis.text.y = element_text(face = "bold", size = 10),
      axis.title = element_blank(),
      axis.ticks.x = element_blank(),
      plot.margin = unit(c(0, 1, 0, 0), "cm"),
      plot.title = element_text(hjust = 0.5, face = "bold", size = 10),
      strip.text.x = element_blank(),
      strip.text.y = element_blank(),
      legend.title = element_text(face = "bold"),
      legend.text = element_text(face = "bold", size = 10))

plot.Num_reads

plot.failed_libs <- tabReads %>%
  group_by(Kit, Soil_type) %>%
  summarise(failed_lib = sum(NumReads < 10E03)) %>%
  ungroup() %>%
  ggplot(aes(x = Kit, y = failed_lib, fill = Soil_type)) + 
  geom_bar(stat = 'identity', alpha = 0.5, color = "black") +
  facet_wrap(facets = vars(Kit), nrow = 1, scales = "free_x") +
  labs(x = "",
       y = "",
       title = "Number of failed libraries",
       fill = "Sample type") +
  scale_fill_manual(values = colors) +
  theme_bw(base_size = 10) +
  theme(legend.position = "none",
      axis.text.x = element_blank(),
      axis.text.y = element_text(face = "bold", size = 10),
      axis.title = element_blank(),
      axis.ticks.x = element_blank(),
      plot.margin = unit(c(0, 1, 0, 0), "cm"),
      plot.title = element_text(hjust = 0.5, face = "bold", size = 10),
      strip.text.x = element_text(size = 7, face = "bold"),
      strip.text.y = element_text(size = 7, face = "bold"),
      legend.title = element_text(face = "bold"),
      legend.text = element_text(face = "bold", size = 10))

plot.rarecurve <- dat_types %>%
  amp_rarecurve(facet_by = "Kit", color_by = "Soil_type") + 
  facet_grid(cols = vars(Kit)) + 
  scale_color_manual(values = alpha(colors, 0.5)) +
  theme_bw(base_size = 10) +
  labs(x = "",
       y = "",
       title = "Number of observed ASVs as function of sequencing depth") +
  theme_bw(base_size = 10) +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5, face = "bold", size = 10),
        axis.text.x = element_text(angle = 90, face = "bold", vjust = 0.5, size = 10),
        axis.text.y = element_text(face = "bold", size = 10),
        plot.margin = unit(c(1, 1, 0, 0), "cm"),
        strip.background = element_blank(),
        strip.text.x = element_blank())

# Stacked plot
plot.failed_libs / plot.Num_reads / plot.rarecurve
```

**Conclusion: Especially the MIDAS HT kit struggles with humic substance removal**


