---
title: "Scaledown Metagenomes"
author: "TBNJ"
date: "2023-03-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo    = F,
  # fig.pos = "H",
  message = FALSE,
  warning = FALSE
)

library(here)
setwd(here("Scaledown"))

source(here("Scaledown", "scripts/loaddata.R"))
source(here("Scaledown", "scripts/functions.R"))
source(here("Scaledown", "scripts/article_theme.R"))


library(tidyverse)
library(ampvis2)
library(vegan)
library(patchwork)
library(ggpubr)
library(DESeq2)
library(rlist)

colors <- c("grey", "darkgoldenrod2", "royalblue4", "seagreen", "salmon")
rm("metadata.amplicon", "ampvis.amplicon")
```

## Filter Data
```{r remove blanks and pcr controls}
dat.types.meta <- ampvis.metagenome %>%
  amp_subset_samples(!is.na(project_id) & !soil_type %in% c("Blank", "Activated sludge"))

# Create subsets
dat.types.sub3000.meta <- dat.types.meta %>% amp_subset_samples(., minreads = 3000)

rarefy <- min(colSums(dat.types.sub3000.meta[["abund"]]))

set.seed(123)
dat.types.sub3000.meta.ra <- dat.types.sub3000.meta %>%
  amp_subset_samples(., rarefy = rarefy)

dat.types.sub3000.meta.ra.rel_0.1 <- dat.types.sub3000.meta.ra %>%
  amp_subset_samples(normalise = TRUE) %>%
  filter_otus(0.1)

dat.types.sub3000.meta.rel_0.1 <- dat.types.sub3000.meta %>%
  amp_subset_samples(normalise = TRUE) %>%
  filter_otus(0.1)

soil <- unique(dat.types.sub3000.meta$metadata$soil_type) %>% sort(.) %>% as.factor(.)
```

### Bray Curtis variation - metagenomes
# Only OTUs above 0.1 % relative abundance is used
```{r calculate bray curtis variation}
meta_rarecurve <- dat.types.sub3000.meta %>%
  amp_rarecurve(color_by = "lib_volume",
                facet_by = "soil_type",
                facet_scales = "free_x") +
  scale_color_manual(values = c(colors[5], colors[3]))

tmp <- t(dat.types.sub3000.meta.ra.rel_0.1$abund)

bray_distance <- as.matrix(vegan::vegdist(vegan::decostand(tmp, method = "hellinger"))) %>%
    as.data.frame() %>%
    rownames_to_column("seq_id")

metadata <- dat.types.sub3000.meta.ra.rel_0.1$metadata %>%
  select(seq_id, lib_volume, soil_type)

metadata_renamed <- metadata %>%
  dplyr::rename(
    comparison = seq_id,
    compared_soil = soil_type,
    compared_protocol = lib_volume
  )

bray_with_metadata <- metadata %>%
  left_join(bray_distance) %>%
  pivot_longer(!seq_id:soil_type, names_to = "comparison", values_to = "value") %>%
  left_join(metadata_renamed) %>%
  filter(value != 0) %>%
  distinct(value, .keep_all = TRUE)

bray_variation_protocol <- bray_with_metadata %>%
  filter(seq_id != comparison) %>%
  filter(paste0(lib_volume, soil_type) == paste0(compared_protocol, compared_soil)) %>%
  filter(!duplicated(paste0(pmax(seq_id, comparison), pmin(seq_id, comparison)))) %>%
  group_by(soil_type, lib_volume) %>%
  summarise(
    mean_bray = format(round(mean(value, na.rm = TRUE), 2), nsmall = 2),
    sd_bray = format(round(sd(value, na.rm = TRUE), 2), nsmall = 2), .groups = "drop")

bray_variation_between_protocols <- bray_with_metadata %>%
  filter(seq_id != comparison, lib_volume != compared_protocol) %>%
  filter(soil_type == compared_soil) %>%
  filter(!duplicated(paste0(pmax(seq_id, comparison), pmin(seq_id, comparison)))) %>%
  group_by(soil_type) %>%
  summarise(
    mean_bray = format(round(mean(value, na.rm = TRUE), 2), nsmall = 2),
    sd_bray = format(round(sd(value, na.rm = TRUE), 2), nsmall = 2),
    bray_between = paste0(mean_bray, " (", sd_bray, ")")
  ) %>%
  mutate(lib_volume = "Full-scale") %>%
  select(soil_type, lib_volume, bray_between)
```

### Shannon Diversity variation - metagenomes
```{r calculate shannon variation}
tab_reads <- data.frame(num_reads = apply(dat.types.sub3000.meta$abund, 2, sum)) %>%
  rownames_to_column(var = "seq_id")

lab.metadata.metagenome <- metadata.metagenome %>%
  filter(!soil_type %in% c("Blank", "Activated sludge", "PCR-POS", "PCR-NEG")) %>%
  select(seq_id, soil_type, lib_volume, lib_conc)

rarefied_alphadiversity <- dat.types.sub3000.meta.ra %>%
  amp_alphadiv(measure = c("observed", "shannon", "simpson"),
               richness = T) %>%
    dplyr::rename(shannon = Shannon,
                  observed_otu = ObservedOTUs) %>%
  select(seq_id, soil_type, lib_volume, observed_otu, shannon) %>%
  left_join(tab_reads)

seq_meta_sub <- seq_meta %>%
  select(seq_id, before_total_reads, after_total_reads, before_total_bases, after_total_bases) %>%
  filter(!str_detect(seq_id, "PCR")) %>%
  mutate(across(seq_id, ~str_remove(., "-MJ362.*")),
         across(before_total_reads:after_total_reads, ~./(2*1E06)),
         across(before_total_bases:after_total_bases, ~./1E9)) %>%
  dplyr::rename(before_MPE = before_total_reads,
                after_MPE = after_total_reads,
                before_Gbp = before_total_bases,
                after_Gbp = after_total_bases) %>%
  select(-before_MPE, -after_MPE)
  
combine_alpha_beta_with_metadata <- rarefied_alphadiversity %>%
  # add library_concentration
  right_join(lab.metadata.metagenome) %>%
  right_join(seq_meta_sub) %>%
  # Calculate failed libraries
  group_by(soil_type, lib_volume) %>%
  mutate(libs = sum(!is.na(shannon))) %>%
  # Calculate mean and sd for all numeric varaibles.
  summarise(across(where(is.numeric), .fns = list(mean = ~ mean(., na.rm = T), 
                                                  sd = ~ sd(., na.rm = T)), .names = "{.fn}_{.col}")) %>%
  # Format mean and sd
  ungroup() %>%
  mutate(libs = as.character(mean_libs)) %>%
  relocate(c(libs, mean_num_reads, sd_num_reads), .after = lib_volume) %>%
  mutate(
    libs = as.character(mean_libs),
    across(mean_num_reads:sd_num_reads, ~ format(round(., 0))),
    across(mean_observed_otu:sd_observed_otu, ~ format(round(., 0))),
    across(where(is.numeric), ~ format(round(., 2), nsmall = 2))) %>%
  left_join(bray_variation_protocol) %>%
  arrange(soil_type, lib_volume) %>%
  select(-c(sd_libs, mean_libs))
```


## Final table - metagenomes
```{r}
mean_sd_table <- combine_alpha_beta_with_metadata %>%
  pivot_longer(cols = -c(soil_type, lib_volume, libs, contains("sd")), 
               names_to = "mean", values_to = "mean_value", names_prefix = "mean_") %>%
  pivot_longer(cols = -c(soil_type, lib_volume, libs, contains("mean")), 
               names_to = "sd", values_to = "sd_value", names_prefix = "sd_") %>%
  group_by(soil_type, lib_volume) %>%
  # Filter for wrong match
  filter(mean == sd) %>%
  #mutate(across(everything(), ~trimws(.))) %>%
  # Paste mean (sd)
  mutate(mean_sd = paste0(mean_value, " (", sd_value, ")")) %>%
  select(!mean_value:sd_value) %>%
  # make wide and add bray variation between protocols
  pivot_wider(names_from = mean, values_from = mean_sd) %>%
  left_join(bray_variation_between_protocols) %>%
  # Remove duplicate Soil type
  mutate(bray_between = if_else(is.na(bray_between), "", bray_between),
         soil_type = if_else(lib_volume == "Small-scale", "", soil_type),
         lib_volume = if_else(lib_volume == "Small-scale", "1 x 5 µL", "1 x 50 µL")) %>%
  # Relocate columns
  relocate(lib_conc, .after = libs) %>%
  relocate(before_Gbp:after_Gbp, .after = lib_conc)

names <- tibble(
  old_names = colnames(mean_sd_table),
  new_names = c("Soil type", "Protocol", "Libraries", "Library conc. [ng/µL]", "Total Reads [Gbp]", "Quality Trimmed Reads [Gbp]", "16S Reads Count", "Observed OTUs", "Shannon Diversity", "Bray-Curtis Dissimilarity", "BC Dissimilarity\nBetween Protocols"))

final_table <- mean_sd_table %>%
  rename_at(vars(names$old_names), ~ names$new_names)

hjustification_matrix <- as.vector(matrix(c(0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1), ncol = ncol(mean_sd_table), nrow = nrow(mean_sd_table), byrow = TRUE))

xjustification_matrix <- as.vector(matrix(c(0.1, 0.1, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9), ncol = ncol(mean_sd_table), nrow = nrow(mean_sd_table), byrow = TRUE))

alpha_table <- final_table %>%
  ggtexttable(rows = NULL, theme = ttheme(colnames.style = colnames_style(fill = "white"), tbody.style = tbody_style(fill = "white", hjust = hjustification_matrix, x = xjustification_matrix))) %>%
  tab_add_hline(at.row = 1:2, row.side = "top", linewidth = 2) %>%
  tab_add_hline(at.row = seq(4, 10, 2), row.side = "top", linewidth = 1)

alpha_table

ggsave(here("Scaledown", "figures", "table_metagenomes.png"), plot = alpha_table, device = "png", height = 12, width = 19)

writexl::write_xlsx(final_table, here("Scaledown", "figures", "scaledown_metagenomes_table.xlsx"))
```


## Statistical tests - two sample t-test or Mann-Whitney(Wilcoxon) on shannon and bray distance
```{r}
# SHANNON
alpha.all <- rarefied_alphadiversity %>%
  left_join(lab.metadata.metagenome) %>%
  arrange(shannon) %>%
  mutate(shannon_rank = seq(1:length(shannon)))

comp.bray <- bray_with_metadata %>%
  dplyr::rename(bray = value) %>%
  filter(soil_type == compared_soil, lib_volume == compared_protocol) %>%
  arrange(bray) %>%
  mutate(bray_rank = seq(1:length(bray)))

# Test for normality
shapiro.test(alpha.all$shannon)
qqnorm(alpha.all$shannon)
qqline(y = alpha.all$shannon)

bartlett.test(shannon ~ lib_volume, data = alpha.all)
fligner.test(shannon ~ lib_volume, data = alpha.all)
car::leveneTest(shannon ~ lib_volume, data = alpha.all)

bartlett.test(shannon ~ soil_type, data = alpha.all)
fligner.test(shannon ~ soil_type, data = alpha.all)
car::leveneTest(shannon ~ soil_type, data = alpha.all)

shapiro.test(comp.bray$bray)
qqnorm(comp.bray$bray)
qqline(y = comp.bray$bray)

bartlett.test(bray ~ lib_volume, data = comp.bray)
fligner.test(bray ~ lib_volume, data = comp.bray)
car::leveneTest(bray ~ lib_volume, data = comp.bray)

bartlett.test(bray ~ soil_type, data = comp.bray)
fligner.test(bray ~ soil_type, data = comp.bray)
car::leveneTest(bray ~ soil_type, data = comp.bray)

# ANOVA shannon
aov.shannon <- lm(shannon ~ lib_volume * soil_type, data = alpha.all)

aov.shannon %>%
  anova(.) %>%
  data.frame() %>%
  rownames_to_column("Variable") %>%
  mutate(R2 = Sum.Sq / sum(Sum.Sq) * 100) %>%
  mutate(across(R2, ~ as.character(round(., digits = 1)))) %>%
  mutate(pval = ifelse(Pr..F. < 0.001, "< 0.001", round(Pr..F., 2))) %>%
  select(Variable, R2, pval) %>%
  gridExtra::tableGrob(rows = NULL) %>%
  gridExtra::grid.arrange(.)

TukeyHSD(aov(aov.shannon))$lib_volume

data.shannon <- alpha.all %>% filter(!is.na(shannon)) %>% pull(shannon)
residuals.shannon <- resid(aov.shannon)
df.shannon <- cbind(data.shannon, residuals.shannon) %>% as.data.frame(.)

ggplot(aes(x = data.shannon,
           y = residuals.shannon),
       data = df.shannon) +
  geom_point() +
  geom_hline(yintercept = 0)

aov.shannon.ranks <- lm(shannon_rank ~ lib_volume * soil_type, data = alpha.all)

aov.shannon.ranks %>%
  anova(.) %>%
  data.frame() %>%
  rownames_to_column("Variable") %>%
  mutate(R2 = Sum.Sq / sum(Sum.Sq) * 100) %>%
  mutate(across(R2, ~ as.character(round(., digits = 1)))) %>%
  mutate(pval = ifelse(Pr..F. < 0.001, "< 0.001", round(Pr..F., 2))) %>%
  select(Variable, R2, pval) %>%
  gridExtra::tableGrob(rows = NULL) %>%
  gridExtra::grid.arrange(.)

TukeyHSD(aov(aov.shannon.ranks))$lib_volume

data.shannon.rank <- alpha.all %>% filter(!is.na(shannon_rank)) %>% pull(shannon_rank)
residuals.shannon.rank <- resid(aov.shannon.ranks)
df.shannon.rank <- cbind(data.shannon.rank, residuals.shannon.rank) %>% as.data.frame(.)

ggplot(aes(x = data.shannon.rank,
           y = residuals.shannon.rank),
       data = df.shannon.rank) +
  geom_point() +
  geom_hline(yintercept = 0)

# Non-parametric test for differences in Shannon index
shannon.small <- alpha.all %>% filter(lib_volume == "Small-scale") %>% pull(shannon)
shannon.full <- alpha.all %>% filter(lib_volume == "Full-scale") %>% pull(shannon)

test.shannon <- wilcox.test(shannon.small, shannon.full, alternative = "two.sided")
test.shannon$p.value

# ANOVA bray
aov.bray <- lm(bray ~ lib_volume * soil_type, data = comp.bray)

aov.bray %>%
  anova(.) %>%
  data.frame() %>%
  rownames_to_column("Variable") %>%
  mutate(R2 = Sum.Sq / sum(Sum.Sq) * 100) %>%
  mutate(across(R2, ~ as.character(round(., digits = 1)))) %>%
  mutate(pval = ifelse(Pr..F. < 0.001, "< 0.001", round(Pr..F., 2))) %>%
  select(Variable, R2, pval) %>%
  gridExtra::tableGrob(rows = NULL) %>%
  gridExtra::grid.arrange(.)

TukeyHSD(aov(aov.bray))$soil_type
TukeyHSD(aov(aov.bray))$`lib_volume:soil_type`

data.bray <- comp.bray %>% filter(!is.na(bray)) %>% pull(bray)
residuals.bray <- resid(aov.bray)
df.bray <- cbind(data.bray, residuals.bray) %>% as.data.frame(.)

ggplot(aes(x = data.bray,
           y = residuals.bray),
       data = df.bray) +
  geom_point() +
  geom_hline(yintercept = 0)

aov.bray.ranks <- lm(bray_rank ~ lib_volume * soil_type, data = comp.bray)

aov.bray.ranks %>%
  anova(.) %>%
  data.frame() %>%
  rownames_to_column("Variable") %>%
  mutate(R2 = Sum.Sq / sum(Sum.Sq) * 100) %>%
  mutate(across(R2, ~ as.character(round(., digits = 1)))) %>%
  mutate(pval = ifelse(Pr..F. < 0.001, "< 0.001", round(Pr..F., 2))) %>%
  select(Variable, R2, pval) %>%
  gridExtra::tableGrob(rows = NULL) %>%
  gridExtra::grid.arrange(.)

TukeyHSD(aov(aov.bray.ranks))$lib_volume

data.bray.rank <- comp.bray %>% filter(!is.na(bray_rank)) %>% pull(bray_rank)
residuals.bray.rank <- resid(aov.bray.ranks)
df.bray.rank <- cbind(data.bray.rank, residuals.bray.rank) %>% as.data.frame(.)

ggplot(aes(x = data.bray.rank,
           y = residuals.bray.rank),
       data = df.bray.rank) +
  geom_point() +
  geom_hline(yintercept = 0)

# Non-parametric test for differences in Shannon index
bray_protocol <- bray_with_metadata %>%
  dplyr::rename(bray = value) %>%
  filter(bray != 0) %>%
  filter(soil_type == compared_soil, !lib_volume == compared_protocol)

bray.BS <- comp.bray %>% filter(soil_type == "Beach Sand") %>% pull(bray)
shapiro.test(bray.BS)
bray.CL <- comp.bray %>% filter(soil_type == "Clay") %>% pull(bray)
shapiro.test(bray.CL)
bray.OR <- comp.bray %>% filter(soil_type == "Organic") %>% pull(bray)
shapiro.test(bray.OR)
bray.SA <- comp.bray %>% filter(soil_type == "Sand") %>% pull(bray)
shapiro.test(bray.SA)
bray.SC <- comp.bray %>% filter(soil_type == "Sand-Clay") %>% pull(bray)
shapiro.test(bray.SC)

bray.protocol.BS <- bray_protocol %>% filter(soil_type == "Beach Sand") %>% pull(bray)
shapiro.test(bray.protocol.BS)
bray.protocol.CL <- bray_protocol %>% filter(soil_type == "Clay") %>% pull(bray)
shapiro.test(bray.protocol.CL)
bray.protocol.OR <- bray_protocol %>% filter(soil_type == "Organic") %>% pull(bray)
shapiro.test(bray.protocol.OR)
bray.protocol.SA <- bray_protocol %>% filter(soil_type == "Sand") %>% pull(bray)
shapiro.test(bray.protocol.SA)
bray.protocol.SC <- bray_protocol %>% filter(soil_type == "Sand-Clay") %>% pull(bray)
shapiro.test(bray.protocol.SC)

test.bray.BS <- wilcox.test(bray.BS, bray.protocol.BS, alternative = "two.sided")
test.bray.CL <- wilcox.test(bray.CL, bray.protocol.CL, alternative = "two.sided")
test.bray.OR <- wilcox.test(bray.OR, bray.protocol.OR, alternative = "two.sided")
test.bray.SA <- wilcox.test(bray.SA, bray.protocol.SA, alternative = "two.sided")
test.bray.SC <- wilcox.test(bray.SC, bray.protocol.SC, alternative = "two.sided")

p.bray <- c(test.bray.BS$p.value, test.bray.CL$p.value, test.bray.OR$p.value, test.bray.SA$p.value, test.bray.SC$p.value)
p.adj_bray <- p.adjust(p.bray, method = "bonferroni")
p.adj_bray
```

## Heatmap - metagenomes
```{r heatmap}
# Find all genera in amplicon data
textmap.metagenome_genus <- dat.types.sub3000.meta %>%
  #amp_subset_samples(soil_type == "Organic") %>%
  amp_heatmap(group_by = "seq_id", 
              min_abundance = 0.001,
              tax_show = 100000,
              tax_aggregate = "Genus", 
              normalise = FALSE,
              showRemainingTaxa = FALSE, 
              plot_values = TRUE,
              textmap = TRUE) %>%
  rownames_to_column(var = "Genus") %>%
  mutate(across(Genus, ~ str_replace_all(., "^.*OTU.*$", "Unclassified"))) %>%
  group_by(Genus) %>%
  summarise(across(everything(), ~ sum(.))) %>%
  ungroup() %>%
  column_to_rownames(var = "Genus")

#genus.text.meta <- heatmap.metagenome_genus %>%
#  mutate(total = rowSums(.)) %>%
#  filter(!total == 0) %>%
#  select(total) %>%
#  arrange(desc(total)) %>%
#  row.names(.)

#genus.text.renamed <- gsub("Burkholderia-Caballeronia-Paraburkholderia", "BCP", genus.text.amp)

# Load new genus ampvis object
ampvis.sub_genus <- amp_load(otutable = textmap.metagenome_genus,
                             metadata = lab.metadata.metagenome)

#plot.heatmap_genus <- ampvis.sub_genus %>%
#  amp_subset_taxa(tax_vector = c("Unclassified"), normalise = TRUE, remove = TRUE) %>%
#  amp_heatmap(group_by = "lib_volume", 
#              min_abundance = 0.01,
#              facet_by = "soil_type",
#              tax_aggregate = "OTU",
#              normalise = FALSE,
#              tax_show = 25,
#              #tax_show = genus.text.amp[2:26], 
#              #order_y_by = rev(genus.text.amp[c(2:26)]),
#              showRemainingTaxa = FALSE, 
#              plot_na = TRUE,
#              plot_values = FALSE,
#              plot_values_size = 3,
#              round = 2,
#              color_vector = c("white", colors[3]),
#              plot_legendbreaks = c(0.01, 0.1, 1, 10),
#              plot_colorscale = "log10") +
#  scale_x_discrete(limits = c("Small-scale", "Full-scale"), labels = c("2 x 5 µL", "2 x 25 µL")) +
#  #scale_y_discrete(labels = rev(c(genus.text.renamed[c(2:26)]))) +
#  labs(x = "", y = "", title = "Heatmap aggregated at the genus level", 
#       tag = "A") +
#  articletheme +
#  guides(fill = guide_colorbar(title.position = "top")) 
#  #theme(plot.margin = unit(c(0, 1, 0, 0), "cm"))

textmap.sub_genus <- ampvis.sub_genus %>%
  amp_subset_taxa(tax_vector = c("Unclassified"), normalise = TRUE, remove = TRUE) %>%
  amp_heatmap(group_by = "lib_volume", 
              min_abundance = 0.01,
              facet_by = "soil_type",
              tax_aggregate = "OTU",
              normalise = FALSE,
              tax_show = 25,
              showRemainingTaxa = FALSE, 
              plot_na = TRUE,
              plot_values = FALSE,
              plot_values_size = 3,
              round = 2,
              color_vector = c("white", colors[3]),
              plot_legendbreaks = c(0.01, 0.1, 1, 10),
              plot_colorscale = "log10",
              textmap = TRUE) %>%
  rownames_to_column(var = "Genus") %>%
  mutate(across(Genus, ~ str_replace_all(., "^.*OTU.*$", "Unclassified"))) %>%
  group_by(Genus) %>%
  summarise(across(everything(), ~ sum(.))) %>%
  ungroup() %>%
  mutate(across(Genus, ~str_replace(., "Burkholderia-Caballeronia-Paraburkholderia", "BCP"))) %>%
  column_to_rownames(var = "Genus")

deseq_results_genus <- tibble(soil = soil) %>%
  mutate(deseq = map(.x = soil, .f = ~ deseq_function(.x, ampvis.sub_genus)))

deseq_results_dataframe_genus <- rlist::list.rbind(deseq_results_genus$deseq) %>%
  mutate(pval_signif = if_else(padj < 0.05, TRUE, FALSE),
         fc_signif = if_else(abs(log2FoldChange) > 1, TRUE, FALSE),)

dat_types_filtered_genus <- ampvis.sub_genus %>%
  amp_subset_samples(normalise = TRUE)

otu_genus <- dat_types_filtered_genus[["abund"]]

ProtocolDifferentialAbundance_genus <- combine_otu_w_metadata(metadata = ampvis.sub_genus$metadata, otu = otu_genus) %>%
  summarise_triplicates() %>%
  calculate_relative_abundance() %>%
  filter_double_zeros() %>%
  left_join(deseq_results_dataframe_genus) %>%
  #filter(sum(`Full-scale`, `Small-scale`) > 0.1) %>%
  mutate(Bias = if_else(log2FoldChange < 0 & pval_signif, "More abundant with 2 x 25 µL", 
                        if_else(pval_signif & log2FoldChange > 0, "More abundant with 2 x 5 µL", "Equally detected")),
         max_abundance = pmax(`Small-scale`, `Full-scale`)) %>%
  filter(!is.na(Bias) && !Bias == "Equally detected")

heatmap.genus_data <- textmap.sub_genus %>%
  rownames_to_column(var = "OTU") %>%
  group_by(OTU) %>%
  pivot_longer(!OTU, names_to = "protocol", values_to = "abund") %>%
  ungroup() %>%
  separate(protocol, into = c("lib_volume", "soil_type"), sep =  " ") %>%
  mutate(across(soil_type, ~str_replace(., "Beach", "Beach Sand"))) %>%
  left_join(ProtocolDifferentialAbundance_genus %>% select(soil_type, OTU, Bias, pval_signif, fc_signif)) %>%
  mutate(across(OTU, ~str_replace(., "Burkholderia-Caballeronia-Paraburkholderia", "BCP"))) %>%
  #mutate(label = if_else(!is.na(Bias), "*", ""))
  mutate(label = case_when(#lib_volume == "Small-scale" & !is.na(Bias) ~ "*",
                           lib_volume == "Full-scale" & Bias == "More abundant with 2 x 25 µL" ~ "*",
                           lib_volume == "Small-scale" & Bias == "More abundant with 2 x 5 µL" ~ "*",
                           TRUE ~ ""))

plot.heatmap_genus <- heatmap.genus_data %>%
ggplot(aes(x = lib_volume, y = OTU)) + 
  geom_tile(aes(fill = abund), colour = "white", size = 0.5) +
  geom_text(aes(label = label), size = 8, vjust = 0.75) +
  facet_wrap(facets = vars(soil_type), ncol = 5) +
  articletheme +
  scale_fill_gradient(low = "white", high = colors[3], trans = "log10", breaks = c(10, 1, 0.1, 0.01, 0.001), na.value = "white") +
  scale_x_discrete(limits = c("Small-scale", "Full-scale"), labels = c("2 x 5 µL", "2 x 25 µL")) +
  scale_y_discrete(limits=rev) +
  #scale_y_discrete(labels = rev(c(genus.text.renamed[c(2:26)]))) +
  labs(x = "", y = "", title = "Heatmap aggregated at the genus level", 
       tag = "A") +
  articletheme +
  guides(fill = guide_colorbar(title.position = "top", title = "% Read Abundance")) 
  #theme(plot.margin = unit(c(0, 1, 0, 0), "cm"))

#plot.heatmap_OTU <- dat.types.sub3000.meta %>%
#  amp_subset_samples(!soil_type == "Beach Sand") %>%
#  amp_heatmap(group_by = "lib_volume", 
#              min_abundance = 0.1,
#              facet_by = "soil_type",
#              tax_aggregate = "OTU",
#              normalise = TRUE,
#              tax_add = c("Phylum"), 
#              tax_show = 25,
#              #tax_show = genus.text.meta[2:26], 
#              #order_y_by = rev(genus.text.meta[c(2:26)]),
#              showRemainingTaxa = FALSE, 
#              plot_na = TRUE,
#              plot_values = FALSE,
#              plot_values_size = 3,
#              round = 2,
#              color_vector = c("white", colors[3]),
#              #plot_legendbreaks = c(0.01, 0.1, 1, 10),
#              plot_colorscale = "sqrt") +
#  scale_x_discrete(limits = c("Small-scale", "Full-scale"), labels = c("2 x 5 µL", "2 x 25 µL")) +
#  #scale_y_discrete(labels = rev(c(genus.text.renamed[c(2:26)]))) +
#  labs(x = "", y = "", title = "Heatmap aggregated at the OTU level", tag = "A") +
#  articletheme +
#  guides(fill = guide_colorbar(title.position = "top"))

leg.plot_heat <- cowplot::get_legend(plot.heatmap_genus)

plot.heatmap_genus.final <- plot.heatmap_genus + theme(legend.position = "none")

plot.heatmap_genus.final

ggsave(here("Scaledown", "figures", "heatmap_metagenomes.png"), plot = plot.heatmap_genus, device = "png", height = 12, width = 19)
```


# PERMANOVA test for overall difference in observed communities based on protocol
```{r statistical testing}
## Overall comparison
permanova.list <- list()
      
for (i in 1:length(soil)) {
  
  community_0.1 <- dat.types.sub3000.meta %>%
    amp_subset_samples(normalise = TRUE) %>%
    filter_otus(0.1) %>%
    amp_subset_samples(soil_type == soil[i])
  
  community_0.1 <- t(community_0.1[["abund"]])
  
  meta <- dat.types.sub3000.meta[["metadata"]] %>%
    filter(soil_type == soil[i])
  
  dis <- vegan::vegdist(community_0.1)
  
  groups.stratified <- meta %>%
    mutate(group = str_c(soil_type, lib_volume, sep = " ")) %>%
    pull(group) %>%
    as.factor(.)
  
 permanova <- adonis2(community_0.1 ~ lib_volume, type = "bray", data = meta, permutations = 99999) %>%
   as.data.frame() %>%
   rownames_to_column("Variable") %>%
   mutate(across(R2, ~ as.character(round(. * 100, digits = 1)))) %>%
   mutate(pval = ifelse(`Pr(>F)` < 0.001, "< 0.001", round(`Pr(>F)`, 2))) %>%
   select(Variable, R2, pval)
 
 permanova.list[[i]] <- permanova
 
 names(permanova.list) <- soil[i]
}
  
names(permanova.list) <- soil

perm.results <- as.numeric(c(permanova.list[[1]][1,2], permanova.list[[2]][1,2], permanova.list[[3]][1,2], permanova.list[[4]][1,2], permanova.list[[5]][1,2]))

shapiro.test(perm.results)
mean(perm.results)
sd(perm.results)

# Statified PERMANOVA
# Betadispersion - analysis of multivariate homogeneity of group dispersions (variances)
community_0.1 <- dat.types.sub3000.meta %>%
  amp_subset_samples(normalise = TRUE) %>%
  filter_otus(0.1)

community_0.1 <- t(community_0.1[["abund"]])

meta <- dat.types.sub3000.meta[["metadata"]]

dis <- vegan::vegdist(community_0.1)

groups.stratified <- meta %>%
  mutate(group = str_c(soil_type, lib_volume, sep = " ")) %>%
  pull(group) %>%
  as.factor(.)

mod.stratified <- vegan::betadisper(dis, groups.stratified)
mod.stratified

anova(mod.stratified)

plot(mod.stratified, ellipse = TRUE, hull = FALSE, conf = 0.90) # 90% data ellipse

perm.strat <- with(meta, permute::how(nperm = 99999, blocks = soil_type))

set.seed(123)
permanova.stratified <- adonis2(community_0.1 ~ lib_volume*soil_type, type = "bray", data = meta, permutations = 99999)

permanova.stratified %>%
  data.frame() %>%
  rownames_to_column("Variable") %>%
  mutate(across(R2, ~ as.character(round(. * 100, digits = 1)))) %>%
  mutate(pval = ifelse(Pr..F. < 0.001, "< 0.001", round(Pr..F., 2))) %>%
  select(Variable, R2, pval) %>%
  gridExtra::tableGrob(rows = NULL) %>%
  gridExtra::grid.arrange(.)
```


## Ordination - metagenomes
```{r ordination}
plot.ordinate <- dat.types.sub3000.meta %>%
  amp_ordinate(filter_species = 0.1,
               sample_shape_by = "lib_volume",
               #sample_color_by = "soil_type",
               type = "PCA",
               transform = "hellinger",
               distmeasure = "bray",
               species_nlabels = 10,
               species_label_taxonomy = "Phylum",
               species_plot = T) +
  ggforce::geom_mark_ellipse(aes(fill = soil_type, group = soil_type), alpha = 0.5, color = "black") +
  scale_color_manual(values = colors[c(1,3,4,5,2)]) +
  scale_fill_manual(values = colors[c(1,3,4,5,2)]) +
  scale_shape_manual(values = c(16,21), labels = c("2 x 25 µL", "2 x 5 µL")) +
  labs(title = "All samples: PCA of microbial diversity",
       fill = "Soil type",
       shape = "Library volume",
       tag = "A",
       subtitle = paste0("**PERMANOVA:**", 
                         "<br>Library volume: Variance explained = ","0.3 %,", " p-value = 0.17",
                         "<br>Soil type: Variance explained = ","95.8 %,", " p-value < 0.001",
                         "<br>Interaction: Variance explained = ","0.7 %,", " p-value = 0.36")) +
  coord_cartesian(xlim = c(-0.75, 0.75), ylim = c(-0.75, 0.75)) +
  articlethemex0 +
  theme(plot.subtitle = ggtext::element_markdown())

plot.ordinate

# Per soil type
plot_ord1 <- dat.types.sub3000.meta %>%
  amp_subset_samples(soil_type == soil[1]) %>%
  amp_ordinate(
    filter_species = 0.1,
    sample_shape_by = "lib_volume",
    type = "PCA",
    transform = "hellinger",
    distmeasure = "bray",
    species_nlabels = 10,
    species_label_taxonomy = "Phylum",
    species_plot = T) +
  ggforce::geom_mark_ellipse(aes(fill = soil_type, group = lib_volume), alpha = 0.5, color = "black") +
  scale_fill_manual(values = colors[1]) +
  scale_shape_manual(values = c(16,21), labels = c("2 x 25 µL", "2 x 5 µL")) +
  labs(title = paste0(soil[1], ": ", "PCA of microbial diversity"),
       fill = "Soil type",
       shape = "Library volume", 
       tag = "B",
       subtitle = paste0("**PERMANOVA:**", "<br>Library volume: Variance explained = ", permanova.list[[1]][1,2], " %,", " p-value = ", permanova.list[[1]][1,3], "<br>", "<br>")) +
  coord_cartesian(xlim = c(-0.75, 0.75), ylim = c(-0.75, 0.75)) +
  articlethemex0 +
  theme(legend.position = "none",
        plot.subtitle = ggtext::element_markdown())
  
plot_ord2 <- dat.types.sub3000.meta %>%
  amp_subset_samples(soil_type == soil[2]) %>%
  amp_ordinate(
    filter_species = 0.1,
    sample_shape_by = "lib_volume",
    type = "PCA",
    transform = "hellinger",
    distmeasure = "bray",
    species_nlabels = 10,
    species_label_taxonomy = "Phylum",
    species_plot = T) +
  ggforce::geom_mark_ellipse(aes(fill = soil_type, group = lib_volume), alpha = 0.5, color = "black") +
  scale_fill_manual(values = colors[3]) +
  scale_shape_manual(values = c(16,21), labels = c("2 x 25 µL", "2 x 5 µL")) +
  labs(title = paste0(soil[2], ": ", "PCA of microbial diversity"),
       fill = "Soil type",
       shape = "Library volume", 
       tag = "C",
       subtitle = paste0("**PERMANOVA:**", "<br>Library volume: Variance explained = ", permanova.list[[2]][1,2], " %,", " p-value = ", permanova.list[[2]][1,3], "<br>", "<br>")) +
  coord_cartesian(xlim = c(-0.75, 0.75), ylim = c(-0.75, 0.75)) +
  articlethemex0 +
  theme(legend.position = "none",
        plot.subtitle = ggtext::element_markdown())

plot_ord3 <- dat.types.sub3000.meta %>%
  amp_subset_samples(soil_type == soil[3]) %>%
  amp_ordinate(
    filter_species = 0.1,
    sample_shape_by = "lib_volume",
    type = "PCA",
    transform = "hellinger",
    distmeasure = "bray",
    species_nlabels = 10,
    species_label_taxonomy = "Phylum",
    species_plot = T) +
  ggforce::geom_mark_ellipse(aes(fill = soil_type, group = lib_volume), alpha = 0.5, color = "black") +
  scale_fill_manual(values = colors[4]) +
  scale_shape_manual(values = c(16,21), labels = c("2 x 25 µL", "2 x 5 µL")) +
  labs(title = paste0(soil[3], ": ", "PCA of microbial diversity"),
       fill = "Soil type",
       shape = "Library volume", 
       tag = "D",
       subtitle = paste0("**PERMANOVA:**", "<br>Library volume: Variance explained = ", permanova.list[[3]][1,2], " %,", " p-value = ", permanova.list[[3]][1,3], "<br>", "<br>")) +
  coord_cartesian(xlim = c(-0.75, 0.75), ylim = c(-0.75, 0.75)) +
  articlethemex0 +
  theme(legend.position = "none",
        plot.subtitle = ggtext::element_markdown())

plot_ord4 <- dat.types.sub3000.meta %>%
  amp_subset_samples(soil_type == soil[4]) %>%
  amp_ordinate(
    filter_species = 0.1,
    sample_shape_by = "lib_volume",
    type = "PCA",
    transform = "hellinger",
    distmeasure = "bray",
    species_nlabels = 10,
    species_label_taxonomy = "Phylum",
    species_plot = T) +
  ggforce::geom_mark_ellipse(aes(fill = soil_type, group = lib_volume), alpha = 0.5, color = "black") +
  scale_fill_manual(values = colors[5]) +
  scale_shape_manual(values = c(16,21), labels = c("2 x 25 µL", "2 x 5 µL")) +
  labs(title = paste0(soil[4], ": ", "PCA of microbial diversity"),
       fill = "Soil type",
       shape = "Library volume", 
       tag = "E",
       subtitle = paste0("**PERMANOVA:**", "<br>Library volume: Variance explained = ", permanova.list[[4]][1,2], " %,", " p-value = ", permanova.list[[4]][1,3], "<br>", "<br>")) +
  coord_cartesian(xlim = c(-0.75, 0.75), ylim = c(-0.75, 0.75)) +
  articlethemex0 +
  theme(legend.position = "none",
        plot.subtitle = ggtext::element_markdown())

plot_ord5 <- dat.types.sub3000.meta %>%
  amp_subset_samples(soil_type == soil[5]) %>%
  amp_ordinate(
    filter_species = 0.1,
    sample_shape_by = "lib_volume",
    type = "PCA",
    transform = "hellinger",
    distmeasure = "bray",
    species_nlabels = 10,
    species_label_taxonomy = "Phylum",
    species_plot = T) +
  ggforce::geom_mark_ellipse(aes(fill = soil_type, group = lib_volume), alpha = 0.5, color = "black") +
  scale_fill_manual(values = colors[2]) +
  scale_shape_manual(values = c(16,21), labels = c("2 x 25 µL", "2 x 5 µL")) +
  labs(title = paste0(soil[5], ": ", "PCA of microbial diversity"),
       fill = "Soil type",
       shape = "Library volume", 
       tag = "F",
       subtitle = paste0("**PERMANOVA:**", "<br>Library volume: Variance explained = ", permanova.list[[5]][1,2], " %,", " p-value = ", permanova.list[[5]][1,3], "<br>", "<br>")) +
  coord_cartesian(xlim = c(-0.75, 0.75), ylim = c(-0.75, 0.75)) +
  articlethemex0 +
  theme(legend.position = "none",
        plot.subtitle = ggtext::element_markdown())

leg.plot.ord_shape <- cowplot::get_legend(plot.ordinate + guides(fill = "none"))

leg.plot.ord_fill <- cowplot::get_legend(plot.ordinate + guides(shape = "none"))

plot_ord_leg <- cowplot::plot_grid(leg.plot.ord_fill, leg.plot.ord_shape, ncol = 1, align = "v", nrow = 2, rel_heights = c(1, 1))

plot.ord <- ggarrange(plot.ordinate, plot_ord1, plot_ord2, plot_ord3, plot_ord4, plot_ord5, common.legend = TRUE, ncol = 3, nrow = 2, legend = "top")

plot.ord

plot_ord_final <- plot_ord3 +
  labs(fill = "Soil type",
       tag = "C",
       shape = "Library volume", 
       subtitle = paste0("**PERMANOVA:**", "<br>Library volume: Variance explained = ", permanova.list[[3]][1,2], " %,", " p-value = ", permanova.list[[3]][1,3])) +
  theme(legend.position = "none")

plot.ordinate.final <- plot.ordinate +
  labs(fill = "Soil type",
       tag = "C",
       shape = "Library volume") +
  theme(legend.position = "none")

ggsave(here("Scaledown", "figures", "PCA_faceted_metagenomes.png"), plot = plot.ord, device = "png", height = 12, width = 19)
```


## Rel abundance correlation plot
# Next chunk merges replicates and sample type. It needs to be one plot per soil type
```{r differential relative abundance}
deseq_results <- tibble(soil = soil) %>%
  mutate(deseq = map(.x = soil, .f = ~ deseq_function(.x, dat.types.sub3000.meta)))

deseq_results_dataframe <- rlist::list.rbind(deseq_results$deseq) %>%
  mutate(pval_signif = if_else(padj < 0.05, TRUE, FALSE),
         fc_signif = if_else(abs(log2FoldChange) > 1, TRUE, FALSE),)

dat_types_filtered <- dat.types.sub3000.meta %>%
  amp_subset_samples(normalise = TRUE)
  #filter_otus(0.1)

otu <- dat_types_filtered[["abund"]]

ProtocolDifferentialAbundance <- combine_otu_w_metadata(metadata = dat.types.sub3000.meta$metadata, otu = otu) %>%
  summarise_triplicates() %>%
  calculate_relative_abundance() %>%
  filter_double_zeros() %>%
  left_join(deseq_results_dataframe) %>%
  #filter(sum(`Full-scale`, `Small-scale`) > 0.1) %>%
  mutate(Bias = if_else(log2FoldChange < 0 & pval_signif, "More abundant with 1 x 50 µL", 
                        if_else(pval_signif & log2FoldChange > 0, "More abundant with 1 x 5 µL", "Equally detected")),
         max_abundance = pmax(`Small-scale`, `Full-scale`)) %>%
  filter(!is.na(Bias))

points.ns <- ProtocolDifferentialAbundance %>%
  filter(Bias == "Equally detected")

points.signf <- ProtocolDifferentialAbundance %>%
  filter(!Bias == "Equally detected")

plot.diff <- ggplot() +
  geom_point(data = points.ns, aes(x = `Full-scale`, y = `Small-scale`, fill = Bias), alpha = 0.2, 
             color = "black", size = 4, position = position_jitter(width = 0.001, height = 0.001), shape = 21) +
  geom_point(data = points.signf, aes(x = `Full-scale`, y = `Small-scale`, fill = Bias), alpha = 0.8, 
             color = "black", size = 4, position = position_jitter(width = 0.001, height = 0.001), shape = 21) +
  coord_cartesian(xlim = c(0, max(ProtocolDifferentialAbundance$max_abundance)), ylim = c(0, max(ProtocolDifferentialAbundance$max_abundance))) +
  scale_fill_manual(values = c("grey", colors[5], colors[3]), 
                    limits = c("Equally detected",
                               "More abundant with 1 x 50 µL", 
                               "More abundant with 1 x 5 µL")) +
  geom_abline(linewidth = 0.5) +
  articlethemex0 +
  labs(title = "All samples: Differential abundance of ASVs",
       x = "Average relative abundance within replicates [%], 1 x 50 µL",
       y = "Average relative abundance within replicates [%], 1 x 5 µL")

plot.diff

plot_list <- map(.x = soil, ~ plotDiffRelabund_meta(ProtocolDifferentialAbundance, .x)) 

ggleg <- plot_list[[3]] + theme(legend.position = "right") 

leg <- list(get_legend(ggleg))

plot_diff <- plot.diff + 
  labs(tag = "B") +
  #labs(title = "Organic: Differential abundance of ASVs", subtitle =  "ASVs with relative abundance above 0.1 %", tag = "B") + 
  #coord_cartesian(xlim = c(0, 0.2), ylim = c(0, 0.2)) +
  theme(legend.position = "none")

plot_list2 <- append(plot_list, list(plot.diff), after = 0)

p.diff <- ggarrange(plotlist = plot_list2, common.legend = TRUE, ncol = 3, nrow = 2, legend = "top", labels = toupper(letters[1:6]), font.label = list(size = 18, face = "bold"))

p.diff

ggsave(here("Scaledown", "figures", "differential_abundance_faceted_metagenomes.png"), plot = p.diff, device = "png", height = 12, width = 19)
```

Note on p-values set to NA: some values in the results table can be set to NA for one of the following reasons:

* If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p value and adjusted p value will all be set to NA.
* If a row contains a sample with an extreme count outlier then the p value and adjusted p value will be set to NA. These outlier counts are detected by Cook’s distance. Customization of this outlier filtering and description of functionality for replacement of outlier counts and refitting is described below
* If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p value will be set to NA. Description and customization of independent filtering is described below
`NAs can be omitted by setting independentFiltering = FALSE, cooksCutoff = FALSE`


## Final plot

```{r final plot}
# Create combined plot
plot <- ggarrange(plot_diff, plot.ordinate.final, ncol = 1, nrow = 2)

plot

leg.heat <- cowplot::get_legend(plot.heatmap_genus)
leg.ord <- cowplot::get_legend(plot.ordinate + theme(legend.position = "right"))
leg.diff <- cowplot::get_legend(plot.diff + theme(legend.position = "right"))

leg.comb <- cowplot::plot_grid(leg.heat, leg.diff, leg.ord, ncol = 1, align = "v", nrow = 3, rel_heights = c(1, 1, 1))

leg.comb

gg_arranged <- ggarrange(plot.heatmap_genus.final, plot, leg.comb, ncol = 3, nrow = 1, widths = c(1, 1, 0.5))

gg_arranged

ggsave(here("Scaledown", "figures", "combined_plot_metagenomes.png"), plot = gg_arranged, device = "png", height = 12, width = 20)
```





