---
title: "Scaledown Metagenomes and Amplicons comparison"
author: "TBNJ"
date: "2023-03-22"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo    = F,
  # fig.pos = "H",
  message = FALSE,
  warning = FALSE
)

source("scripts/loaddata.R", encoding = "UTF-8")
source("./article_theme.R")
source("./functions.R")

library(tidyverse)
library(ampvis2)
library(patchwork)
library(ggpubr)

colors <- c("grey", "darkgoldenrod2", "royalblue4", "seagreen", "#9D3B38") #,"salmon")
```

## Filter Data
```{r remove blanks and pcr controls}
# Create subsets
dat.types.sub3000.meta <- ampvis.metagenome %>% 
  amp_subset_samples(!is.na(project_id) & !soil_type %in% c("Blank", "Activated sludge"), minreads = 3000)

dat.types.sub3000.meta.ra <- dat.types.sub3000.meta %>%
  amp_rarefy(rarefy = min(colSums(dat.types.sub3000.meta[["abund"]])))

# Create subsets
dat.types.sub3000.amp <- ampvis.amplicon %>%
  amp_subset_samples(!is.na(project_id) & !soil_type %in% c("Blank", "Activated sludge"), minreads = 3000) %>%
  amp_subset_samples(!seq_id == "MFD004-4-F3")

dat.types.sub3000.amp.ra <- dat.types.sub3000.amp %>%
  amp_rarefy(rarefy = min(colSums(dat.types.sub3000.amp[["abund"]])))

## Metadata
metadata.metagenome <- metadata.metagenome %>%
  filter(!is.na(project_id) & !soil_type %in% c("Blank", "Activated sludge")) %>%
  mutate(across(seq_id, ~str_replace(., "MFD004", "Metagenome")))

metadata.amplicon <- metadata.amplicon %>%
  filter(!is.na(project_id) & !soil_type %in% c("Blank", "Activated sludge")) %>%
  mutate(across(seq_id, ~str_replace(., "MFD004", "Amplicon")))
```

## Tax assignments
TO DO: Summarise into one table together with seq-stats
```{r}
# Summarise classifications on each tax level
tax.amp <- dat.types.sub3000.amp[["tax"]] %>%
  select(-OTU, -Species) %>%
  mutate(across(everything(), ~na_if(., "")))

tax.meta <- dat.types.sub3000.meta[["tax"]] %>%
  select(-OTU, -Species) %>%
  mutate(across(everything(), ~na_if(., "")))

# Create stats for percent annotated reads at each tax level
stats.annotated.amp <- tax.amp %>%
  #filter(Kingdom == "k:Bacteria") %>%
  #select(-Kingdom) %>%
  summarise(across(.cols = everything(), .fns = num_annotated, .names = "{col}")) %>%
  pivot_longer(cols = 1:6, names_to = "Tax_level", values_to = "Assigned") %>%
  mutate(Tax_level = factor(Tax_level, unique(Tax_level))) %>%
  arrange(desc(Assigned)) %>%
  mutate(across(Tax_level, ~as.factor(.))) %>%
  mutate(Unassigned = max(Assigned)-Assigned) %>%
  pivot_longer(cols = 2:3, names_to = "Assigned", values_to = "Count") %>%
  mutate(Percent = Count/max(Count)*100)

stats.annotated.meta <- tax.meta %>%
  #filter(Kingdom == "k:Bacteria") %>%
  #select(-Kingdom) %>%
  summarise(across(.cols = everything(), .fns = num_annotated, .names = "{col}")) %>%
  pivot_longer(cols = 1:6, names_to = "Tax_level", values_to = "Assigned") %>%
  mutate(Tax_level = factor(Tax_level, unique(Tax_level))) %>%
  arrange(desc(Assigned)) %>%
  mutate(across(Tax_level, ~as.factor(.))) %>%
  mutate(Unassigned = max(Assigned)-Assigned) %>%
  pivot_longer(cols = 2:3, names_to = "Assigned", values_to = "Count") %>%
  mutate(Percent = Count/max(Count)*100)

# Create overview of taxa found
tax_summary.amp <- tax.amp %>%
  #filter(Kingdom == "k:Bacteria") %>%
  #select(-Kingdom) %>%
  summarise(across(everything(), ~n_distinct(., na.rm = TRUE))) %>%
  pivot_longer(cols = 1:6, names_to = "Tax_level", values_to = "Count") %>%
  mutate(across(Tax_level, ~factor(., levels = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus"))))

tax_summary.meta <- tax.meta %>%
  #filter(Kingdom == "k:Bacteria") %>%
  #select(-Kingdom) %>%
  summarise(across(everything(), ~n_distinct(., na.rm = TRUE))) %>%
  pivot_longer(cols = 1:6, names_to = "Tax_level", values_to = "Count") %>%
  mutate(across(Tax_level, ~factor(., levels = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus"))))

tax_summary.combined <- tax_summary.amp %>%
  dplyr::rename(Amplicon = Count) %>%
  merge(tax_summary.meta %>% dplyr::rename(Metagenome = Count)) %>%
  arrange(Amplicon)

# Plots
plot1 <- stats.annotated.amp %>%
  ggplot(aes(y = reorder(Tax_level, Count), x = Percent, fill = factor(Assigned, levels = c("Unassigned", "Assigned")))) +
  geom_bar(stat = "identity", alpha = 1, position = "fill") +
  scale_fill_manual(values = colors[c(5, 3)]) +
  scale_y_discrete(limits = rev) +
  theme_bw(base_size = 18) +
  labs(title = "Amplicons", y = "", x = "", fill = "Stats") +
  theme(title = element_text(face = "bold"),
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        axis.title.x = element_text(face = "bold")) +
  articletheme
  
plot2 <- stats.annotated.meta %>%
  ggplot(aes(y = reorder(Tax_level, Count), x = Percent, fill = factor(Assigned, levels = c("Unassigned", "Assigned")))) +
  geom_bar(stat = "identity", alpha = 1, position = "fill") +
  scale_fill_manual(values = colors[c(5, 3)]) +
  scale_y_discrete(limits = rev) +
  theme_bw(base_size = 18) +
  labs(title = "Metagenomes", y = "", x = "", fill = "Stats") +
  theme(title = element_text(face = "bold"),
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        axis.title.x = element_text(face = "bold")) +
  articletheme

plot1 + plot2 + plot_layout(guides = "collect")

rm(tax_summary.amp, tax_summary.meta)
```


## Combine data - Genus
```{r}
# Create un-normalised aggreagated genus-tables
heatmap.amplicon_genus <- dat.types.sub3000.amp %>%
  amp_heatmap(group_by = "seq_id", 
              min_abundance = 0.001,
              tax_show = 100000,
              tax_aggregate = "Genus", 
              normalise = FALSE,
              showRemainingTaxa = FALSE, 
              plot_values = TRUE,
              textmap = TRUE) %>%
  rownames_to_column(var = "Genus") %>%
  mutate(across(Genus, ~ str_replace_all(., "^.*ASV.*$", "Unclassified"))) %>%
  group_by(Genus) %>%
  summarise(across(everything(), ~ sum(.))) %>%
  ungroup() %>%
  #arrange(match(Genus, genus.text.amp)) %>%
  rename_with(., ~str_replace(., "MFD004", "Amplicon")) %>%
  column_to_rownames(var = "Genus")

heatmap.metagenome_genus <- dat.types.sub3000.meta %>%
  amp_heatmap(group_by = "seq_id", 
              min_abundance = 0.001,
              tax_show = 100000,
              tax_aggregate = "Genus", 
              normalise = FALSE,
              showRemainingTaxa = FALSE, 
              plot_values = TRUE,
              textmap = TRUE) %>%
  rownames_to_column(var = "Genus") %>%
  mutate(across(Genus, ~ str_replace_all(., "^.*OTU.*$", "Unclassified"))) %>%
  group_by(Genus) %>%
  summarise(across(everything(), ~ sum(.))) %>%
  ungroup() %>%
  #arrange(match(Genus, genus.text.meta)) %>%
  rename_with(., ~str_replace(., "MFD004", "Metagenome")) %>%
  column_to_rownames(var = "Genus")

# Find all genera in amplicon data
genus.text.amp <- heatmap.amplicon_genus %>%
  mutate(total = rowSums(.)) %>%
  filter(!total == 0) %>%
  select(total) %>%
  arrange(desc(total)) %>%
  row.names(.)

# Find all genera in metagenomic data
genus.text.meta <- heatmap.metagenome_genus %>%
  mutate(total = rowSums(.)) %>%
  filter(!total == 0) %>%
  select(total) %>%
  arrange(desc(total)) %>%
  row.names(.)

# Combine data into one full table
heatmap.combined_genus <- full_join(heatmap.amplicon_genus %>% rownames_to_column(var = "Genus"), 
                                    heatmap.metagenome_genus %>% rownames_to_column(var = "Genus")) %>%
  column_to_rownames(var = "Genus") %>%
  replace(is.na(.), 0)

# Combine metadata into one file
metadata.combined <- rbind(metadata.metagenome, metadata.amplicon) %>%
  mutate(across(experiment, ~str_extract(., "[^_]+$")),
         across(experiment, ~str_to_title(.))) %>%
  mutate(volume_experiment = str_c(lib_volume, experiment, sep = " "),
         type_experiment = str_c(soil_type, experiment, sep = " "))

# Load new combined ampvis object
ampvis.sub.combined_genus <- amp_load(otutable = heatmap.combined_genus,
                                      metadata = metadata.combined)
```

## Heatmap - Genus
```{r}
# Find all genera combined data
genus.text.combined <- heatmap.combined_genus %>%
  mutate(total = rowSums(.)) %>%
  filter(!total == 0) %>%
  select(total) %>%
  arrange(desc(total)) %>%
  row.names(.)

plot3 <- ampvis.sub.combined_genus %>%
  amp_heatmap(group_by = "experiment", 
              min_abundance = 0.01,
              facet_by = "soil_type",
              tax_aggregate = "OTU",
              normalise = TRUE,
              tax_show = genus.text.combined[1:49], 
              order_y_by = rev(genus.text.combined[c(2:49, 1)]),
              showRemainingTaxa = TRUE, 
              plot_na = FALSE,
              plot_values = TRUE,
              plot_values_size = 3,
              round = 2,
              color_vector = c(colors[5], "white", colors[3]),
              plot_colorscale = "log10",
              plot_legendbreaks = c(0.01, 1, 50)) +
  scale_x_discrete(limits = c("Amplicon", "Metagenome")) +
  labs(x = "", y = "", title = "Heatmap aggregated at the genus level across data type") +
  articletheme +
  theme(legend.position = "bottom") +
  guides(fill = guide_colorbar(title.position = "right"))
```

## Ordination - Genus
```{r}
plot4 <- ampvis.sub.combined_genus %>%
  amp_subset_samples(., normalise = TRUE) %>%
  #amp_subset_samples(experiment == "Small_v_full_metagenome") %>%
  #amp_subset_samples(soil_type == "Organic") %>%
  amp_ordinate(filter_species = 0.001,
               type = "NMDS",
               distmeasure = "bray",
               transform = "none",
               sample_color_by = "soil_type",
               species_plot = FALSE,
               species_nlabels = 0,
               species_label_taxonomy = "OTU",
               sample_shape_by = "volume_experiment",
               sample_colorframe = FALSE) +
  scale_color_manual(values = colors) +
  scale_fill_manual(values = colors) +
  scale_shape_manual(values = c(0, 15, 1, 16), breaks = c("Full-scale Amplicon", "Small-scale Amplicon", "Full-scale Metagenome", "Small-scale Metagenome")) +
  #ggforce::geom_mark_ellipse(aes(color = soil_type, group = facet), alpha = 0.5) +
  ggforce::geom_mark_ellipse(aes(fill = soil_type, group = type_experiment), alpha = 0.3, color = "black") +
  labs(title = "NMDS at genus level") +
  ## Add factor level sorting
  articlethemex0
```



## Genus overlap of taxa stats
## Stats - alpha diversity, bray distance and PERMANOVA


## Alpha diversity metric plot
# Evaluate per soil type
```{r}
ampvis.sub.combined_genus_filtered_0.001_ra_rel <- ampvis.sub.combined_genus %>%
  filter_otus(0.001) %>%
  amp_subset_samples(., rarefy = min(colSums(.[["abund"]])))

rarefied_alphadiversity <- ampvis.sub.combined_genus_filtered_0.001_ra_rel %>%
  amp_alpha_diversity(measure = c("observed", "shannon", "simpson"), richness = TRUE) %>%
  select(soil_type, experiment, lib_volume, type_experiment, volume_experiment, ObservedOTUs, Shannon, Simpson, Chao1, ACE) %>%
  dplyr::rename(Observed_genera = ObservedOTUs)

tabReads <- merge(
  rarefied_alphadiversity,
  data.frame(NumReads = apply(ampvis.sub.combined_genus$abund, 2, sum)),
  by = 0)

alpha <- tabReads %>%
  group_by(soil_type) %>%
  pivot_longer(Observed_genera:ACE, names_to = "index", values_to = "diversity")
  
plot.alpha1 <- alpha %>%
  filter(index %in% c("Shannon", "Observed_genera", "Chao1", "ACE")) %>%
  ggplot(aes(x = volume_experiment, y = diversity, fill = soil_type)) +
  facet_grid(cols = vars(soil_type), rows = vars(index), scales = "free_y") +
  geom_boxplot(outlier.size = 0, alpha = 0.5) +
  #geom_violin(alpha = 0.5) +
  geom_point() +
  scale_fill_manual(values = colors) +
  scale_x_discrete(limits = c("Full-scale Amplicon", "Small-scale Amplicon", "Full-scale Metagenome", "Small-scale Metagenome")) +
  articletheme +
  labs(title = "")

plot.alpha2 <- alpha %>%
  filter(index %in% c("Shannon", "Observed_genera", "Chao1", "ACE")) %>%
  ggplot(aes(x = experiment, y = diversity)) +
  facet_grid(cols = vars(experiment), rows = vars(index), scales = "free") +
  geom_boxplot(outlier.size = 0, alpha = 0.5) +
  #geom_violin(alpha = 0.5) +
  #geom_point() +
  #scale_fill_manual(values = colors) +
  articletheme +
  labs(title = "")

plot.alpha2 + plot.alpha1
```

### Bray Curtis variation - Metagenomes
# Only OTUs above 0.001 % relative abundance is used
```{r calculate bray curtis variation}
bray_with_metadata <- calculate_bray_distance2(ampvis.sub.combined_genus_filtered_0.001_ra_rel)

bray_variation_protocol <- bray_with_metadata %>%
  filter(seq_id != comparison) %>%
  filter(paste0(experiment, soil_type) == paste0(compared_protocol, compared_soil)) %>%
  filter(!duplicated(paste0(pmax(seq_id, comparison), pmin(seq_id, comparison)))) %>%
  group_by(soil_type, experiment) %>%
  summarise(
    mean_bray = format(round(mean(value, na.rm = TRUE), 2), nsmall = 2),
    sd_bray = format(round(sd(value, na.rm = TRUE), 2), nsmall = 2), .groups = "drop")

bray_variation_between_protocols <- bray_with_metadata %>%
  filter(seq_id != comparison, experiment != compared_protocol) %>%
  filter(soil_type == compared_soil) %>%
  filter(!duplicated(paste0(pmax(seq_id, comparison), pmin(seq_id, comparison)))) %>%
  group_by(soil_type) %>%
  summarise(
    mean_bray = format(round(mean(value, na.rm = TRUE), 2), nsmall = 2),
    sd_bray = format(round(sd(value, na.rm = TRUE), 2), nsmall = 2),
    bray_between = paste0(mean_bray, " (", sd_bray, ")")
  ) %>%
  mutate(experiment = "Metagenome") %>% # What to write here?
  select(soil_type, experiment, bray_between)
```

### Shannon Diversity variation - Metagenomes
```{r calculate shannon variation}
lab.metadata.combined <- metadata.combined %>%
  select(seq_id, soil_type, experiment, lib_conc)

rarefied_alphadiversity2 <- ampvis.sub.combined_genus %>%
  amp_alphadiv(measure = c("observed", "shannon", "simpson"),
               richness = T) %>%
  dplyr::rename(shannon = Shannon) %>%
  select(seq_id, soil_type, experiment, shannon)

combine_alpha_beta_with_metadata <- rarefied_alphadiversity2 %>%
  # add library_concentration
  left_join(lab.metadata.combined) %>%
  # Calculate failed libraries
  group_by(soil_type, experiment) %>%
  mutate(libs = n()) %>%
  # Calculate mean and sd for all numeric varaibles.
  summarise(across(where(is.numeric), .fns = list(mean = ~ mean(., na.rm = T), 
                                                  sd = ~ sd(., na.rm = T)), .names = "{.fn}_{.col}")) %>%
  # Format mean and sd
  ungroup() %>%
  mutate(libs = as.character(mean_libs),
         across(where(is.numeric), ~ format(round(., 2), nsmall = 2))) %>%
  left_join(bray_variation_protocol) %>%
  arrange(soil_type, experiment) %>%
  select(-c(sd_libs, mean_libs))
```


Add PERMANOVA stats
## Final table - Combined
```{r create final table}
mean_sd_table <- combine_alpha_beta_with_metadata %>%
  pivot_longer(cols = -c(soil_type, experiment, libs, contains("sd")), 
               names_to = "mean", values_to = "mean_value", names_prefix = "mean_") %>%
  pivot_longer(cols = -c(soil_type, experiment, libs, contains("mean")), 
               names_to = "sd", values_to = "sd_value", names_prefix = "sd_") %>%
  group_by(soil_type, experiment) %>%
  # Filter for wrong match
  filter(mean == sd) %>%
  # Paste mean (sd)
  mutate(mean_sd = paste0(mean_value, " (", sd_value, ")")) %>%
  select(!mean_value:sd_value) %>%
  # make wide and add bray variation between protocols
  pivot_wider(names_from = mean, values_from = mean_sd) %>%
  left_join(bray_variation_between_protocols) %>%
  # Remove duplicate Soil type
  mutate(bray_between = if_else(is.na(bray_between), "", bray_between)) %>%
  # Relocate columns
  relocate(lib_conc, .after = libs)


names <- tibble(
  old_names = colnames(mean_sd_table),
  new_names = c("Soil type", "Protocol", "Libraries", "Library Conc.\n[ng/ÂµL]", "Shannon\nDiversity", "Bray-Curtis\nVariation", "Bray-Curtis Variation\nBetween protocols"))

final_table <- mean_sd_table %>%
  rename_at(vars(names$old_names), ~ names$new_names)

hjustification_matrix <- as.vector(matrix(c(0, 0, 1, 1, 1, 1, 1), ncol = ncol(mean_sd_table), nrow = nrow(mean_sd_table), byrow = TRUE))

xjustification_matrix <- as.vector(matrix(c(0.1, 0.1, 0.9, 0.9, 0.9, 0.9, 0.9), ncol = ncol(mean_sd_table), nrow = nrow(mean_sd_table), byrow = TRUE))

alpha_table <- final_table %>%
  ggtexttable(rows = NULL, theme = ttheme(colnames.style = colnames_style(fill = "white"), tbody.style = tbody_style(fill = "white", hjust = hjustification_matrix, x = xjustification_matrix))) %>%
  tab_add_hline(at.row = 1:2, row.side = "top", linewidth = 2) %>%
  tab_add_hline(at.row = seq(4, 10, 2), row.side = "top", linewidth = 1)

alpha_table

ggsave(plot = alpha_table, filename = "table_combined.png", device = "png", height = 13, width = 16)
```




## Genera Set differences 
```{r}
genus.text.amp.all <- tax.amp %>%
  select(Genus) %>%
  distinct(.) %>%
  mutate(across(Genus, ~replace_na(., "Unclassified"))) %>%
  pull(Genus)

genus.text.meta.all <- tax.meta %>%
  select(Genus) %>%
  distinct(.) %>%
  mutate(across(Genus, ~replace_na(., "Unclassified"))) %>%
  pull(Genus)

only.amplicon.all <- setdiff(genus.text.amp.all, genus.text.meta.all)
only.amplicon.all %>% length(.)

only.metagenome.all <- setdiff(genus.text.meta.all, genus.text.amp.all)
only.metagenome.all %>% length(.)

intersect.all <- intersect(genus.text.amp.all, genus.text.meta.all)
intersect.all %>% length(.)

agg.data <- ampvis.sub.combined_genus %>%
  amp_heatmap(group_by = "seq_id",
              min_abundance = 0.001,
              tax_aggregate = "OTU",
              measure = "median",
              tax_show = 100000,
              normalise = FALSE,
              textmap = TRUE) %>%
  mutate(total = rowSums(.)) %>%
  filter(!total == 0) %>%
  arrange(desc(total)) %>%
  rownames_to_column(var = "Genus") %>%
  mutate(across(Genus, ~ str_replace_all(., "^.*ASV.*$", "Unclassified"))) %>%
  mutate(across(Genus, ~ str_replace_all(., "^.*OTU.*$", "Unclassified"))) %>%
  group_by(Genus) %>%
  pivot_longer(!Genus, names_to = "seq_id", values_to = "abund") %>%
  merge(metadata.combined %>% select(seq_id, experiment, soil_type, type_experiment))

agg1 <- agg.data %>%
  group_by(Genus, experiment) %>%
  summarise(across(abund, ~sum(.)), .groups = "drop")

agg2 <- agg1 %>%
  group_by(Genus) %>%
  pivot_wider(names_from = "experiment", values_from = "abund") %>%
  column_to_rownames(var = "Genus")

agg3 <- agg.data %>%
  group_by(Genus, type_experiment) %>%
  summarise(across(abund, ~sum(.)), .groups = "drop")

agg4 <- agg3 %>%
  group_by(Genus) %>%
  pivot_wider(names_from = "type_experiment", values_from = "abund") %>%
  column_to_rownames(var = "Genus")

amp.test <- amp_load(otutable = agg2)

amp.test2 <- amp_load(otutable = agg4)

rare1 <- amp.test %>%
  amp_subset_taxa(tax_vector = intersect.all) %>%
  amp_rarecurve(color_by = "SampleID",
                stepsize = 10) +
  geom_hline(yintercept = 292) +
  #geom_hline(yintercept = 263) +
  #geom_hline(yintercept = 277) +
  geom_vline(xintercept = 72500) +
  labs(y = "Number of observed Genera")

rare2 <- amp.test2 %>%
  amp_subset_taxa(tax_vector = intersect.all) %>%
  amp_rarecurve(color_by = "SampleID",
                stepsize = 10) +
  geom_hline(yintercept = 292) +
  #geom_hline(yintercept = 263) +
  #geom_hline(yintercept = 277) +
  geom_vline(xintercept = 72500) +
  labs(y = "Number of observed Genera")

rare2 + rare1 + plot_layout(guides = "collect")




amp.test %>%
  #amp_subset_taxa(tax_vector = intersect.all) %>%
  amp_rarecurve(color_by = "SampleID",
                stepsize = 100) +
  geom_vline(xintercept = 72500) +
  geom_hline(yintercept = 532)



```

## Cost analysis
```{r}
# constant1 = 14700 (metagenomic 16s reads for same genus diversity)
# Number of total metagenomic reads / total metagenomic 16s reads = constant2 (metagenomic reads per 1 meta 16 read)
# 900 * constant2 = Number of needed metagenomic reads for same genus diversity
constant1 <- 72500

meta.reads <- seq_meta %>%
  dplyr::rename(seq_name = seq_id) %>%
  left_join(metadata.metagenome) %>%
  filter(library_id %in% ampvis.sub.combined_genus$metadata$library_id) %>%
  select(before_total_reads) %>%
  mutate(across(before_total_reads, ~./2)) %>% # Change to number of PE reads
  sum(.)

meta16.reads <- tabReads %>%
  filter(experiment == "Metagenome") %>%
  select(NumReads) %>%
  sum(.)

constant2 = meta.reads / meta16.reads

reads.needed = constant1*constant2

reads.needed

reads.needed / 29


# Cost NovaSeq6000
# How many PE reads on a S4 lane (8-10 billion / 4 = 2-2.5 B)
# How many samples possible = 96*4 = 384
# How many reads per sample = 2 or 2.5 / 384
reads.needed.per.sample = reads.needed / nrow(tabReads %>% filter(experiment == "metagenome"))

s4.low = 2E09 / reads.needed.per.sample
s4.high = 2.2E09 / reads.needed.per.sample

fraction.per.lane.low =  reads.needed / 2E09 
fraction.per.lane.high =  reads.needed / 2.2E09

2E09/ 96*3
2.2E09 / 96*3

cost.s4 = (1E05 / 4 * fraction.per.lane.low)
cost.s4.possible = (1E05 / (4*3*96))
cost.total.s4 = cost.s4 + (50 * nrow(tabReads %>% filter(experiment == "metagenome")))
cost.total.s4.possible = cost.s4.possible + 50 + 50


# Cost MiSeq
# How many SE reads on a lane (22-25 million)
# How many samples possible = 96*4 = 384
# How many reads per sample = 22 or 25 / 384
reads.amp <- tabReads %>%
  filter(experiment == "amplicon") %>%
  select(NumReads) %>%
  sum(.)

reads.needed.per.sample.amp = reads.amp / nrow(tabReads %>% filter(experiment == "metagenome"))

miseq.low = 20E06 / reads.needed.per.sample.amp
miseq.high = 25E06 / reads.needed.per.sample.amp

fraction.miseq.low =  reads.amp / 20E06
fraction.miseq.high =  reads.amp / 25E06

cost.miseq = (4E04 * fraction.miseq.low)
cost.miseq.possible = (17E03 / (3*96))
cost.total.miseq = cost.miseq + 20 * nrow(tabReads %>% filter(experiment == "amplicon"))
cost.total.miseq.possible = cost.miseq.possible + 50
```



## Heatmap - Genus
```{r}
# Find all genera combined data
ampvis.sub.combined_genus %>%
  amp_subset_taxa(tax_vector = intersect.all) %>%
  amp_heatmap(group_by = "experiment", 
              min_abundance = 0.01,
              facet_by = "soil_type",
              tax_aggregate = "OTU",
              normalise = TRUE,
              tax_show = genus.text.combined[1:49], 
              order_y_by = rev(genus.text.combined[c(2:49, 1)]),
              showRemainingTaxa = TRUE, 
              plot_na = FALSE,
              plot_values = TRUE,
              plot_values_size = 3,
              round = 2,
              color_vector = c(colors[5], "white", colors[3]),
              plot_colorscale = "log10",
              plot_legendbreaks = c(0.01, 1, 50)) +
  #scale_x_discrete(limits = c("Small-scale Amplicon", "Full-scale Amplicon", "Small-scale Metagenome", "Full-scale Metagenome")) +
  scale_x_discrete(limits = c("Amplicon", "Metagenome")) +
  labs(x = "", y = "", title = "Heatmap aggregated at the genus level across data type") +
  articletheme +
  theme(legend.position = "bottom") +
  guides(fill = guide_colorbar(title.position = "right"))
```

## Ordination - Genus Intersect
```{r}
ampvis.sub.combined_genus %>%
  amp_subset_taxa(tax_vector = intersect.all) %>%
  amp_ordinate(filter_species = 0.001,
               type = "NMDS",
               distmeasure = "jaccard",
               transform = "pa",
               sample_color_by = "soil_type",
               species_plot = FALSE,
               species_nlabels = 0,
               species_label_taxonomy = "OTU",
               sample_shape_by = "volume_experiment",
               sample_colorframe = FALSE) +
  scale_color_manual(values = colors) +
  scale_fill_manual(values = colors) +
  scale_shape_manual(values = c(0, 15, 1, 16), breaks = c("Full-scale Amplicon", "Small-scale Amplicon", "Full-scale Metagenome", "Small-scale Metagenome")) +
  #ggforce::geom_mark_ellipse(aes(color = soil_type, group = facet), alpha = 0.5) +
  ggforce::geom_mark_ellipse(aes(fill = soil_type, group = type_experiment), alpha = 0.3, color = "black") +
  labs(title = "NMDS at genus level") +
  ## Add factor level sorting
  articlethemex0
```


